{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "from metapaths.tools.processing import regularize_colnames, head, expand_col_on_char, combine_group_rows_on_char\n",
    "from hetnet_ml.src import graph_tools as gt\n",
    "\n",
    "load_dir = Path('../2_pipeline/00_download_data/out/').resolve()\n",
    "network_dir = load_dir.parent.parent.joinpath('01_Building_a_network_from_CTD/out')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_nodes = gt.remove_colons(pd.read_csv(network_dir.joinpath('nodes_all.csv'), dtype=str))\n",
    "edges = gt.remove_colons(pd.read_csv(network_dir.joinpath('edges.csv'), dtype=str))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "abv, et = gt.get_abbrev_dict_and_edge_tuples(all_nodes, edges)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "go_ids = set(all_nodes[all_nodes['id'].str.startswith('GO:')]['id'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_nodes = []\n",
    "new_edges = []"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Start loading .gaf files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "go_cols = ['db', 'db_object_id', 'db_object_symbol','qualifier', 'go_id', 'db_reference', 'evidence_code',\n",
    " 'with_or_from', 'aspect', 'db_object_name', 'db_object_synonym', 'db_object_type', 'taxon', 'date', 'assigned_by',\n",
    " 'annotation_extension', 'gene_product_form_id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>db</th>\n",
       "      <th>db_object_id</th>\n",
       "      <th>db_object_symbol</th>\n",
       "      <th>qualifier</th>\n",
       "      <th>go_id</th>\n",
       "      <th>db_reference</th>\n",
       "      <th>evidence_code</th>\n",
       "      <th>with_or_from</th>\n",
       "      <th>aspect</th>\n",
       "      <th>db_object_name</th>\n",
       "      <th>db_object_synonym</th>\n",
       "      <th>db_object_type</th>\n",
       "      <th>taxon</th>\n",
       "      <th>date</th>\n",
       "      <th>assigned_by</th>\n",
       "      <th>annotation_extension</th>\n",
       "      <th>gene_product_form_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>A0A024R161</td>\n",
       "      <td>DNAJC25-GNG10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>GO_REF:0000002</td>\n",
       "      <td>IEA</td>\n",
       "      <td>InterPro:IPR001770</td>\n",
       "      <td>F</td>\n",
       "      <td>Guanine nucleotide-binding protein subunit gamma</td>\n",
       "      <td>DNAJC25-GNG10|hCG_1994888</td>\n",
       "      <td>protein</td>\n",
       "      <td>taxon:9606</td>\n",
       "      <td>20190504</td>\n",
       "      <td>InterPro</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>A0A024RBG1</td>\n",
       "      <td>NUDT4B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0003723</td>\n",
       "      <td>GO_REF:0000037</td>\n",
       "      <td>IEA</td>\n",
       "      <td>UniProtKB-KW:KW-0694</td>\n",
       "      <td>F</td>\n",
       "      <td>Diphosphoinositol polyphosphate phosphohydrola...</td>\n",
       "      <td>NUDT4B</td>\n",
       "      <td>protein</td>\n",
       "      <td>taxon:9606</td>\n",
       "      <td>20190504</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          db db_object_id db_object_symbol qualifier       go_id  \\\n",
       "0  UniProtKB   A0A024R161    DNAJC25-GNG10       NaN  GO:0003924   \n",
       "1  UniProtKB   A0A024RBG1           NUDT4B       NaN  GO:0003723   \n",
       "\n",
       "     db_reference evidence_code          with_or_from aspect  \\\n",
       "0  GO_REF:0000002           IEA    InterPro:IPR001770      F   \n",
       "1  GO_REF:0000037           IEA  UniProtKB-KW:KW-0694      F   \n",
       "\n",
       "                                      db_object_name  \\\n",
       "0   Guanine nucleotide-binding protein subunit gamma   \n",
       "1  Diphosphoinositol polyphosphate phosphohydrola...   \n",
       "\n",
       "           db_object_synonym db_object_type       taxon      date assigned_by  \\\n",
       "0  DNAJC25-GNG10|hCG_1994888        protein  taxon:9606  20190504    InterPro   \n",
       "1                     NUDT4B        protein  taxon:9606  20190504     UniProt   \n",
       "\n",
       "  annotation_extension gene_product_form_id  \n",
       "0                  NaN                  NaN  \n",
       "1                  NaN                  NaN  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_prot = pd.read_csv(load_dir.joinpath('goa_human.gaf.gz'), sep='\\t', header=None, \n",
    "                      names=go_cols, comment='!', dtype=str)\n",
    "go_prot.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "UniProtKB    479438\n",
       "Name: db, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_prot['db'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>db</th>\n",
       "      <th>db_object_id</th>\n",
       "      <th>db_object_symbol</th>\n",
       "      <th>qualifier</th>\n",
       "      <th>go_id</th>\n",
       "      <th>db_reference</th>\n",
       "      <th>evidence_code</th>\n",
       "      <th>with_or_from</th>\n",
       "      <th>aspect</th>\n",
       "      <th>db_object_name</th>\n",
       "      <th>db_object_synonym</th>\n",
       "      <th>db_object_type</th>\n",
       "      <th>taxon</th>\n",
       "      <th>date</th>\n",
       "      <th>assigned_by</th>\n",
       "      <th>annotation_extension</th>\n",
       "      <th>gene_product_form_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ComplexPortal</td>\n",
       "      <td>CPX-1012</td>\n",
       "      <td>tenascin-w_human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0030155</td>\n",
       "      <td>PMID:17909022</td>\n",
       "      <td>IDA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P</td>\n",
       "      <td>Tenascin-W complex</td>\n",
       "      <td>TNN hexamer|Tenascin W complex</td>\n",
       "      <td>protein_complex</td>\n",
       "      <td>taxon:9606</td>\n",
       "      <td>20170313</td>\n",
       "      <td>ComplexPortal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ComplexPortal</td>\n",
       "      <td>CPX-1012</td>\n",
       "      <td>tenascin-w_human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0030334</td>\n",
       "      <td>PMID:17909022</td>\n",
       "      <td>IDA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P</td>\n",
       "      <td>Tenascin-W complex</td>\n",
       "      <td>TNN hexamer|Tenascin W complex</td>\n",
       "      <td>protein_complex</td>\n",
       "      <td>taxon:9606</td>\n",
       "      <td>20170313</td>\n",
       "      <td>ComplexPortal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              db db_object_id  db_object_symbol qualifier       go_id  \\\n",
       "0  ComplexPortal     CPX-1012  tenascin-w_human       NaN  GO:0030155   \n",
       "1  ComplexPortal     CPX-1012  tenascin-w_human       NaN  GO:0030334   \n",
       "\n",
       "    db_reference evidence_code with_or_from aspect      db_object_name  \\\n",
       "0  PMID:17909022           IDA          NaN      P  Tenascin-W complex   \n",
       "1  PMID:17909022           IDA          NaN      P  Tenascin-W complex   \n",
       "\n",
       "                db_object_synonym   db_object_type       taxon      date  \\\n",
       "0  TNN hexamer|Tenascin W complex  protein_complex  taxon:9606  20170313   \n",
       "1  TNN hexamer|Tenascin W complex  protein_complex  taxon:9606  20170313   \n",
       "\n",
       "     assigned_by annotation_extension gene_product_form_id  \n",
       "0  ComplexPortal                  NaN                  NaN  \n",
       "1  ComplexPortal                  NaN                  NaN  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_complex = pd.read_csv(load_dir.joinpath('goa_human_complex.gaf.gz'), sep='\\t', header=None, \n",
    "                      names=go_cols, dtype=str, comment='!')\n",
    "go_complex.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>db</th>\n",
       "      <th>db_object_id</th>\n",
       "      <th>db_object_symbol</th>\n",
       "      <th>qualifier</th>\n",
       "      <th>go_id</th>\n",
       "      <th>db_reference</th>\n",
       "      <th>evidence_code</th>\n",
       "      <th>with_or_from</th>\n",
       "      <th>aspect</th>\n",
       "      <th>db_object_name</th>\n",
       "      <th>db_object_synonym</th>\n",
       "      <th>db_object_type</th>\n",
       "      <th>taxon</th>\n",
       "      <th>date</th>\n",
       "      <th>assigned_by</th>\n",
       "      <th>annotation_extension</th>\n",
       "      <th>gene_product_form_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ComplexPortal</td>\n",
       "      <td>CPX-1012</td>\n",
       "      <td>tenascin-w_human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0062023</td>\n",
       "      <td>PMID:19884327</td>\n",
       "      <td>IDA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>Tenascin-W complex</td>\n",
       "      <td>TNN hexamer|Tenascin W complex</td>\n",
       "      <td>protein_complex</td>\n",
       "      <td>taxon:9606</td>\n",
       "      <td>20170313</td>\n",
       "      <td>ComplexPortal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ComplexPortal</td>\n",
       "      <td>CPX-1032</td>\n",
       "      <td>snurportin_human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0005654</td>\n",
       "      <td>PMID:9670026</td>\n",
       "      <td>IDA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>Importin complex, Snurportin variant</td>\n",
       "      <td>Snurportin complex|Importin complex, SNUPN var...</td>\n",
       "      <td>protein_complex</td>\n",
       "      <td>taxon:9606</td>\n",
       "      <td>20170608</td>\n",
       "      <td>ComplexPortal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               db db_object_id  db_object_symbol qualifier       go_id  \\\n",
       "2   ComplexPortal     CPX-1012  tenascin-w_human       NaN  GO:0062023   \n",
       "13  ComplexPortal     CPX-1032  snurportin_human       NaN  GO:0005654   \n",
       "\n",
       "     db_reference evidence_code with_or_from aspect  \\\n",
       "2   PMID:19884327           IDA          NaN      C   \n",
       "13   PMID:9670026           IDA          NaN      C   \n",
       "\n",
       "                          db_object_name  \\\n",
       "2                     Tenascin-W complex   \n",
       "13  Importin complex, Snurportin variant   \n",
       "\n",
       "                                    db_object_synonym   db_object_type  \\\n",
       "2                      TNN hexamer|Tenascin W complex  protein_complex   \n",
       "13  Snurportin complex|Importin complex, SNUPN var...  protein_complex   \n",
       "\n",
       "         taxon      date    assigned_by annotation_extension  \\\n",
       "2   taxon:9606  20170313  ComplexPortal                  NaN   \n",
       "13  taxon:9606  20170608  ComplexPortal                  NaN   \n",
       "\n",
       "   gene_product_form_id  \n",
       "2                   NaN  \n",
       "13                  NaN  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_complex.query('aspect == \"C\"').head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ComplexPortal    1470\n",
       "Name: db, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_complex['db'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>db</th>\n",
       "      <th>db_object_id</th>\n",
       "      <th>db_object_symbol</th>\n",
       "      <th>qualifier</th>\n",
       "      <th>go_id</th>\n",
       "      <th>db_reference</th>\n",
       "      <th>evidence_code</th>\n",
       "      <th>with_or_from</th>\n",
       "      <th>aspect</th>\n",
       "      <th>db_object_name</th>\n",
       "      <th>db_object_synonym</th>\n",
       "      <th>db_object_type</th>\n",
       "      <th>taxon</th>\n",
       "      <th>date</th>\n",
       "      <th>assigned_by</th>\n",
       "      <th>annotation_extension</th>\n",
       "      <th>gene_product_form_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>A0A087WTH5</td>\n",
       "      <td>KCNE1B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0005249</td>\n",
       "      <td>GO_REF:0000002</td>\n",
       "      <td>IEA</td>\n",
       "      <td>InterPro:IPR000369|InterPro:IPR005424</td>\n",
       "      <td>F</td>\n",
       "      <td>Potassium voltage-gated channel subfamily E me...</td>\n",
       "      <td>KCNE1B</td>\n",
       "      <td>protein</td>\n",
       "      <td>taxon:9606</td>\n",
       "      <td>20190504</td>\n",
       "      <td>InterPro</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UniProtKB:A0A087WU88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>A0A087WTH5</td>\n",
       "      <td>KCNE1B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0005249</td>\n",
       "      <td>GO_REF:0000002</td>\n",
       "      <td>IEA</td>\n",
       "      <td>InterPro:IPR000369|InterPro:IPR005424</td>\n",
       "      <td>F</td>\n",
       "      <td>Potassium voltage-gated channel subfamily E me...</td>\n",
       "      <td>KCNE1B</td>\n",
       "      <td>protein</td>\n",
       "      <td>taxon:9606</td>\n",
       "      <td>20190504</td>\n",
       "      <td>InterPro</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UniProtKB:A0A087WWU3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          db db_object_id db_object_symbol qualifier       go_id  \\\n",
       "0  UniProtKB   A0A087WTH5           KCNE1B       NaN  GO:0005249   \n",
       "1  UniProtKB   A0A087WTH5           KCNE1B       NaN  GO:0005249   \n",
       "\n",
       "     db_reference evidence_code                           with_or_from aspect  \\\n",
       "0  GO_REF:0000002           IEA  InterPro:IPR000369|InterPro:IPR005424      F   \n",
       "1  GO_REF:0000002           IEA  InterPro:IPR000369|InterPro:IPR005424      F   \n",
       "\n",
       "                                      db_object_name db_object_synonym  \\\n",
       "0  Potassium voltage-gated channel subfamily E me...            KCNE1B   \n",
       "1  Potassium voltage-gated channel subfamily E me...            KCNE1B   \n",
       "\n",
       "  db_object_type       taxon      date assigned_by annotation_extension  \\\n",
       "0        protein  taxon:9606  20190504    InterPro                  NaN   \n",
       "1        protein  taxon:9606  20190504    InterPro                  NaN   \n",
       "\n",
       "   gene_product_form_id  \n",
       "0  UniProtKB:A0A087WU88  \n",
       "1  UniProtKB:A0A087WWU3  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_isoform = pd.read_csv(load_dir.joinpath('goa_human_isoform.gaf.gz'), sep='\\t', header=None, \n",
    "                      names=go_cols, dtype=str, comment='!')\n",
    "go_isoform.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "UniProtKB    101596\n",
       "Name: db, dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_isoform['db'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>db</th>\n",
       "      <th>db_object_id</th>\n",
       "      <th>db_object_symbol</th>\n",
       "      <th>qualifier</th>\n",
       "      <th>go_id</th>\n",
       "      <th>db_reference</th>\n",
       "      <th>evidence_code</th>\n",
       "      <th>with_or_from</th>\n",
       "      <th>aspect</th>\n",
       "      <th>db_object_name</th>\n",
       "      <th>db_object_synonym</th>\n",
       "      <th>db_object_type</th>\n",
       "      <th>taxon</th>\n",
       "      <th>date</th>\n",
       "      <th>assigned_by</th>\n",
       "      <th>annotation_extension</th>\n",
       "      <th>gene_product_form_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>URS0000001346_9606</td>\n",
       "      <td>URS0000001346_9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0006412</td>\n",
       "      <td>GO_REF:0000108</td>\n",
       "      <td>IEA</td>\n",
       "      <td>GO:0030533</td>\n",
       "      <td>P</td>\n",
       "      <td>Homo sapiens (human) tRNA-Lys</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>taxon:9606</td>\n",
       "      <td>20190504</td>\n",
       "      <td>GOC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>URS0000001346_9606</td>\n",
       "      <td>URS0000001346_9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0030533</td>\n",
       "      <td>GO_REF:0000115</td>\n",
       "      <td>IEA</td>\n",
       "      <td>Rfam:RF00005</td>\n",
       "      <td>F</td>\n",
       "      <td>Homo sapiens (human) tRNA-Lys</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>taxon:9606</td>\n",
       "      <td>20190504</td>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           db        db_object_id    db_object_symbol qualifier       go_id  \\\n",
       "0  RNAcentral  URS0000001346_9606  URS0000001346_9606       NaN  GO:0006412   \n",
       "1  RNAcentral  URS0000001346_9606  URS0000001346_9606       NaN  GO:0030533   \n",
       "\n",
       "     db_reference evidence_code  with_or_from aspect  \\\n",
       "0  GO_REF:0000108           IEA    GO:0030533      P   \n",
       "1  GO_REF:0000115           IEA  Rfam:RF00005      F   \n",
       "\n",
       "                  db_object_name db_object_synonym db_object_type       taxon  \\\n",
       "0  Homo sapiens (human) tRNA-Lys               NaN           tRNA  taxon:9606   \n",
       "1  Homo sapiens (human) tRNA-Lys               NaN           tRNA  taxon:9606   \n",
       "\n",
       "       date assigned_by annotation_extension gene_product_form_id  \n",
       "0  20190504         GOC                  NaN                  NaN  \n",
       "1  20190504  RNAcentral                  NaN                  NaN  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_rna = pd.read_csv(load_dir.joinpath('goa_human_rna.gaf.gz'), sep='\\t', header=None, \n",
    "                      names=go_cols, dtype=str, comment='!')\n",
    "go_rna.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "RNAcentral    43498\n",
       "Name: db, dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_rna['db'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add GO to protein annotiaions....\n",
    "\n",
    "Will also be able to add complex and rna info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "type_conversion = {'P': 'involved_in_GinBP',\n",
    "                   'F': 'enables_GeMF',\n",
    "                   'C': 'part_of_GpoCC'}\n",
    "\n",
    "edges_go_prot = go_prot.rename(columns={'db_object_symbol': 'start_id', 'go_id': 'end_id'})\n",
    "edges_go_prot['type'] = edges_go_prot['aspect'].map(type_conversion)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fix_col_order_edge(df):\n",
    "    cols = ['start_id', 'end_id', 'type']\n",
    "    cols = cols + [c for c in df.columns if c not in cols]\n",
    "    return cols\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_id</th>\n",
       "      <th>end_id</th>\n",
       "      <th>type</th>\n",
       "      <th>db</th>\n",
       "      <th>db_object_id</th>\n",
       "      <th>qualifier</th>\n",
       "      <th>db_reference</th>\n",
       "      <th>evidence_code</th>\n",
       "      <th>with_or_from</th>\n",
       "      <th>aspect</th>\n",
       "      <th>db_object_name</th>\n",
       "      <th>db_object_synonym</th>\n",
       "      <th>db_object_type</th>\n",
       "      <th>taxon</th>\n",
       "      <th>date</th>\n",
       "      <th>assigned_by</th>\n",
       "      <th>annotation_extension</th>\n",
       "      <th>gene_product_form_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DNAJC25-GNG10</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>enables_GeMF</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>A0A024R161</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO_REF:0000002</td>\n",
       "      <td>IEA</td>\n",
       "      <td>InterPro:IPR001770</td>\n",
       "      <td>F</td>\n",
       "      <td>Guanine nucleotide-binding protein subunit gamma</td>\n",
       "      <td>DNAJC25-GNG10|hCG_1994888</td>\n",
       "      <td>protein</td>\n",
       "      <td>taxon:9606</td>\n",
       "      <td>20190504</td>\n",
       "      <td>InterPro</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NUDT4B</td>\n",
       "      <td>GO:0003723</td>\n",
       "      <td>enables_GeMF</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>A0A024RBG1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO_REF:0000037</td>\n",
       "      <td>IEA</td>\n",
       "      <td>UniProtKB-KW:KW-0694</td>\n",
       "      <td>F</td>\n",
       "      <td>Diphosphoinositol polyphosphate phosphohydrola...</td>\n",
       "      <td>NUDT4B</td>\n",
       "      <td>protein</td>\n",
       "      <td>taxon:9606</td>\n",
       "      <td>20190504</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        start_id      end_id          type         db db_object_id qualifier  \\\n",
       "0  DNAJC25-GNG10  GO:0003924  enables_GeMF  UniProtKB   A0A024R161       NaN   \n",
       "1         NUDT4B  GO:0003723  enables_GeMF  UniProtKB   A0A024RBG1       NaN   \n",
       "\n",
       "     db_reference evidence_code          with_or_from aspect  \\\n",
       "0  GO_REF:0000002           IEA    InterPro:IPR001770      F   \n",
       "1  GO_REF:0000037           IEA  UniProtKB-KW:KW-0694      F   \n",
       "\n",
       "                                      db_object_name  \\\n",
       "0   Guanine nucleotide-binding protein subunit gamma   \n",
       "1  Diphosphoinositol polyphosphate phosphohydrola...   \n",
       "\n",
       "           db_object_synonym db_object_type       taxon      date assigned_by  \\\n",
       "0  DNAJC25-GNG10|hCG_1994888        protein  taxon:9606  20190504    InterPro   \n",
       "1                     NUDT4B        protein  taxon:9606  20190504     UniProt   \n",
       "\n",
       "  annotation_extension gene_product_form_id  \n",
       "0                  NaN                  NaN  \n",
       "1                  NaN                  NaN  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges_go_prot = edges_go_prot[fix_col_order_edge(edges_go_prot)]\n",
    "edges_go_prot.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove relations that have qualifiers that contain NOT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "edges_go_prot = edges_go_prot[~edges_go_prot['qualifier'].str.contains('NOT').fillna(False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GeneIDs with more than one symbol in same row: 0\n",
      "Gene Symbols mapped to multiple GeneIDs: 0\n"
     ]
    }
   ],
   "source": [
    "# if 0 then no Pipe Characters in symnbol, so all only have 1 associated symbol\n",
    "print('GeneIDs with more than one symbol in same row: {:,}'.format(all_nodes['gene_symbol']\n",
    "                                                                   .str\n",
    "                                                                   .contains('|', regex=False)\n",
    "                                                                   .sum()))\n",
    "# if 0 then all GeneID to Symbol mappings are 1 to 1\n",
    "print('Gene Symbols mapped to multiple GeneIDs: {:,}'.format((all_nodes.dropna(subset=['gene_symbol'])\n",
    "                                                                       .groupby('gene_symbol')['id']\n",
    "                                                                       .nunique() > 1).sum()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_sym_to_id = all_nodes.set_index('gene_symbol')['id'].dropna().to_dict()\n",
    "edges_go_prot['start_id'] = edges_go_prot['start_id'].map(gene_sym_to_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GO Gene annotations\n",
      "Total:      478,227\n",
      "Mapped:     475,214\n",
      "Unmapped:     3,013\n"
     ]
    }
   ],
   "source": [
    "print('GO Gene annotations')\n",
    "print('Total:   {:10,}'.format(len(edges_go_prot)))\n",
    "print('Mapped:  {:10,}'.format(edges_go_prot.query('end_id in @go_ids')['start_id'].count()))\n",
    "print('Unmapped:{:10,}'.format(len(edges_go_prot) - edges_go_prot.query('end_id in @go_ids')['start_id'].count()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19834"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges_go_prot['db_object_id'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19095"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges_go_prot['start_id'].nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are fewer Gene IDs than Protein IDs, so there may be 1 to many relationships going on..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of UniProt IDs that map to more than one Gene ID: 0\n",
      "Number of Gene IDs that map to more than 1 UniProt ID: 51\n"
     ]
    }
   ],
   "source": [
    "print('Number of UniProt IDs that map to more than one Gene ID: {:,}'.format(\n",
    "    (edges_go_prot[['start_id', 'db_object_id']].groupby('db_object_id')['start_id'].nunique() > 1).sum()))\n",
    "print('Number of Gene IDs that map to more than 1 UniProt ID: {:,}'.format(\n",
    "    (edges_go_prot[['start_id', 'db_object_id']].groupby('start_id')['db_object_id'].nunique() > 1).sum()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "51 Genes that map to multiple Uniprots... it should be fine to map them to all the edges for all the correspoding uniprots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we will do a pipe, separated mapping and add to the node Infos.\n",
    "gene_to_uniprot = combine_group_rows_on_char(edges_go_prot[['start_id', 'db_object_id']], 'start_id', \n",
    "                                             ['db_object_id'], '|')\n",
    "gene_to_uniprot = gene_to_uniprot.set_index('start_id')['db_object_id'].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>label</th>\n",
       "      <th>tree_numbers</th>\n",
       "      <th>drug_bank_ids</th>\n",
       "      <th>alt_disease_ids</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>alt_gene_ids</th>\n",
       "      <th>bio_gridids</th>\n",
       "      <th>pharm_gkbids</th>\n",
       "      <th>uni_prot_ids</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>199609</th>\n",
       "      <td>1</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>100064369|100354232|100400383|100438958|100482...</td>\n",
       "      <td>106523</td>\n",
       "      <td>PA24356</td>\n",
       "      <td>A0A0A0MX79|A0A1U8C678|A0A2J8JM56|A0A2K5C3B8|A0...</td>\n",
       "      <td>P04217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199611</th>\n",
       "      <td>29974</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>100013290|100071844|100155074|100174508|100231...</td>\n",
       "      <td>119004</td>\n",
       "      <td>PA162375098</td>\n",
       "      <td>A0A024QZI9|A0A024QZJ5|A0A024QZM7|A0A091CWX6|A0...</td>\n",
       "      <td>Q9NQ94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199618</th>\n",
       "      <td>2</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A2M</td>\n",
       "      <td>100061692|100173946|100390764|100543551|100657...</td>\n",
       "      <td>106524|231245|246347|679499</td>\n",
       "      <td>PA24357</td>\n",
       "      <td>A0A1U7TC46|A0A2K5E7U5|A0A2K5KBI2|A0A2K6N6A9|A0...</td>\n",
       "      <td>P01023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199622</th>\n",
       "      <td>144568</td>\n",
       "      <td>alpha-2-macroglobulin like 1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>100061421|100127688|100152492|100347314|100407...</td>\n",
       "      <td>126860</td>\n",
       "      <td>PA142670460</td>\n",
       "      <td>A0A0D9RAZ7|A0A1S3AMH8|A0A1U7S0T5|A0A1U7SWH3|A0...</td>\n",
       "      <td>A8K2U0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200727</th>\n",
       "      <td>127550</td>\n",
       "      <td>alpha 1,3-galactosyltransferase 2</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>100339429|100414894|100431438|100479995|100586...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PA142670461</td>\n",
       "      <td>A0A087Y8Z2|A0A1S2ZRA6|A0A1S3FHK7|A0A1U7R263|A0...</td>\n",
       "      <td>U3KPV4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201201</th>\n",
       "      <td>53947</td>\n",
       "      <td>alpha 1,4-galactosyltransferase (P blood group)</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>100017110|100090950|100172807|100347783|100399...</td>\n",
       "      <td>119825</td>\n",
       "      <td>PA143485570|PA24359</td>\n",
       "      <td>A0A096N8Q2|A0A0D9SD56|A0A0S2Z5J1|A0A1S2ZRQ3|A0...</td>\n",
       "      <td>Q9NPC4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201203</th>\n",
       "      <td>51146</td>\n",
       "      <td>alpha-1,4-N-acetylglucosaminyltransferase</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A4GNT</td>\n",
       "      <td>100024221|100033872|100091228|100224446|100351...</td>\n",
       "      <td>119330</td>\n",
       "      <td>PA134960042</td>\n",
       "      <td>A0A091DKX1|A0A096MVI4|A0A0D9REY6|A0A1S3APW6|A0...</td>\n",
       "      <td>Q9UNA3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204630</th>\n",
       "      <td>8086</td>\n",
       "      <td>aladin WD repeat nucleoporin</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>100063811|100154333|100218436|100356621|100405...</td>\n",
       "      <td>113759|230214</td>\n",
       "      <td>PA24361</td>\n",
       "      <td>A0A087XX44|A0A091CTK4|A0A096NFX9|A0A0P7XRK7|A0...</td>\n",
       "      <td>Q9NRG9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204635</th>\n",
       "      <td>65985</td>\n",
       "      <td>acetoacetyl-CoA synthetase</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AACS</td>\n",
       "      <td>100020461|100061661|100076424|100156545|100231...</td>\n",
       "      <td>122434|219689</td>\n",
       "      <td>PA134940696</td>\n",
       "      <td>A0A024RBV2|A0A087XT43|A0A0D9S5D0|A0A0F8CR28|A0...</td>\n",
       "      <td>Q86V21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204640</th>\n",
       "      <td>13</td>\n",
       "      <td>arylacetamide deacetylase</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AADAC</td>\n",
       "      <td>100056411|100148912|100217836|100393374|100468...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PA24363</td>\n",
       "      <td>A0A087QIL4|A0A091QEJ0|A0A091SA32|A0A093J0F4|A0...</td>\n",
       "      <td>P22760</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            id                                             name label  \\\n",
       "199609       1                           alpha-1-B glycoprotein  Gene   \n",
       "199611   29974                   APOBEC1 complementation factor  Gene   \n",
       "199618       2                            alpha-2-macroglobulin  Gene   \n",
       "199622  144568                     alpha-2-macroglobulin like 1  Gene   \n",
       "200727  127550                alpha 1,3-galactosyltransferase 2  Gene   \n",
       "201201   53947  alpha 1,4-galactosyltransferase (P blood group)  Gene   \n",
       "201203   51146        alpha-1,4-N-acetylglucosaminyltransferase  Gene   \n",
       "204630    8086                     aladin WD repeat nucleoporin  Gene   \n",
       "204635   65985                       acetoacetyl-CoA synthetase  Gene   \n",
       "204640      13                        arylacetamide deacetylase  Gene   \n",
       "\n",
       "       tree_numbers drug_bank_ids alt_disease_ids gene_symbol  \\\n",
       "199609          NaN           NaN             NaN        A1BG   \n",
       "199611          NaN           NaN             NaN        A1CF   \n",
       "199618          NaN           NaN             NaN         A2M   \n",
       "199622          NaN           NaN             NaN       A2ML1   \n",
       "200727          NaN           NaN             NaN     A3GALT2   \n",
       "201201          NaN           NaN             NaN      A4GALT   \n",
       "201203          NaN           NaN             NaN       A4GNT   \n",
       "204630          NaN           NaN             NaN        AAAS   \n",
       "204635          NaN           NaN             NaN        AACS   \n",
       "204640          NaN           NaN             NaN       AADAC   \n",
       "\n",
       "                                             alt_gene_ids  \\\n",
       "199609  100064369|100354232|100400383|100438958|100482...   \n",
       "199611  100013290|100071844|100155074|100174508|100231...   \n",
       "199618  100061692|100173946|100390764|100543551|100657...   \n",
       "199622  100061421|100127688|100152492|100347314|100407...   \n",
       "200727  100339429|100414894|100431438|100479995|100586...   \n",
       "201201  100017110|100090950|100172807|100347783|100399...   \n",
       "201203  100024221|100033872|100091228|100224446|100351...   \n",
       "204630  100063811|100154333|100218436|100356621|100405...   \n",
       "204635  100020461|100061661|100076424|100156545|100231...   \n",
       "204640  100056411|100148912|100217836|100393374|100468...   \n",
       "\n",
       "                        bio_gridids         pharm_gkbids  \\\n",
       "199609                       106523              PA24356   \n",
       "199611                       119004          PA162375098   \n",
       "199618  106524|231245|246347|679499              PA24357   \n",
       "199622                       126860          PA142670460   \n",
       "200727                          NaN          PA142670461   \n",
       "201201                       119825  PA143485570|PA24359   \n",
       "201203                       119330          PA134960042   \n",
       "204630                113759|230214              PA24361   \n",
       "204635                122434|219689          PA134940696   \n",
       "204640                          NaN              PA24363   \n",
       "\n",
       "                                             uni_prot_ids uniprot_id  \n",
       "199609  A0A0A0MX79|A0A1U8C678|A0A2J8JM56|A0A2K5C3B8|A0...     P04217  \n",
       "199611  A0A024QZI9|A0A024QZJ5|A0A024QZM7|A0A091CWX6|A0...     Q9NQ94  \n",
       "199618  A0A1U7TC46|A0A2K5E7U5|A0A2K5KBI2|A0A2K6N6A9|A0...     P01023  \n",
       "199622  A0A0D9RAZ7|A0A1S3AMH8|A0A1U7S0T5|A0A1U7SWH3|A0...     A8K2U0  \n",
       "200727  A0A087Y8Z2|A0A1S2ZRA6|A0A1S3FHK7|A0A1U7R263|A0...     U3KPV4  \n",
       "201201  A0A096N8Q2|A0A0D9SD56|A0A0S2Z5J1|A0A1S2ZRQ3|A0...     Q9NPC4  \n",
       "201203  A0A091DKX1|A0A096MVI4|A0A0D9REY6|A0A1S3APW6|A0...     Q9UNA3  \n",
       "204630  A0A087XX44|A0A091CTK4|A0A096NFX9|A0A0P7XRK7|A0...     Q9NRG9  \n",
       "204635  A0A024RBV2|A0A087XT43|A0A0D9S5D0|A0A0F8CR28|A0...     Q86V21  \n",
       "204640  A0A087QIL4|A0A091QEJ0|A0A091SA32|A0A093J0F4|A0...     P22760  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_nodes['uniprot_id'] = all_nodes['id'].map(gene_to_uniprot)\n",
    "all_nodes.dropna(subset=['uniprot_id']).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add all mapped edges\n",
    "new_edges.append(edges_go_prot.dropna(subset=['start_id', 'end_id']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Complex to GO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_id</th>\n",
       "      <th>end_id</th>\n",
       "      <th>type</th>\n",
       "      <th>db</th>\n",
       "      <th>db_object_symbol</th>\n",
       "      <th>qualifier</th>\n",
       "      <th>db_reference</th>\n",
       "      <th>evidence_code</th>\n",
       "      <th>with_or_from</th>\n",
       "      <th>aspect</th>\n",
       "      <th>db_object_name</th>\n",
       "      <th>db_object_synonym</th>\n",
       "      <th>db_object_type</th>\n",
       "      <th>taxon</th>\n",
       "      <th>date</th>\n",
       "      <th>assigned_by</th>\n",
       "      <th>annotation_extension</th>\n",
       "      <th>gene_product_form_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CPX-1012</td>\n",
       "      <td>GO:0030155</td>\n",
       "      <td>involved_in_XinBP</td>\n",
       "      <td>ComplexPortal</td>\n",
       "      <td>tenascin-w_human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:17909022</td>\n",
       "      <td>IDA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P</td>\n",
       "      <td>Tenascin-W complex</td>\n",
       "      <td>TNN hexamer|Tenascin W complex</td>\n",
       "      <td>protein_complex</td>\n",
       "      <td>taxon:9606</td>\n",
       "      <td>20170313</td>\n",
       "      <td>ComplexPortal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CPX-1012</td>\n",
       "      <td>GO:0030334</td>\n",
       "      <td>involved_in_XinBP</td>\n",
       "      <td>ComplexPortal</td>\n",
       "      <td>tenascin-w_human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:17909022</td>\n",
       "      <td>IDA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P</td>\n",
       "      <td>Tenascin-W complex</td>\n",
       "      <td>TNN hexamer|Tenascin W complex</td>\n",
       "      <td>protein_complex</td>\n",
       "      <td>taxon:9606</td>\n",
       "      <td>20170313</td>\n",
       "      <td>ComplexPortal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CPX-1012</td>\n",
       "      <td>GO:0062023</td>\n",
       "      <td>part_of_XpoCC</td>\n",
       "      <td>ComplexPortal</td>\n",
       "      <td>tenascin-w_human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:19884327</td>\n",
       "      <td>IDA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>Tenascin-W complex</td>\n",
       "      <td>TNN hexamer|Tenascin W complex</td>\n",
       "      <td>protein_complex</td>\n",
       "      <td>taxon:9606</td>\n",
       "      <td>20170313</td>\n",
       "      <td>ComplexPortal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CPX-1012</td>\n",
       "      <td>GO:1903672</td>\n",
       "      <td>involved_in_XinBP</td>\n",
       "      <td>ComplexPortal</td>\n",
       "      <td>tenascin-w_human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:19884327</td>\n",
       "      <td>IDA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P</td>\n",
       "      <td>Tenascin-W complex</td>\n",
       "      <td>TNN hexamer|Tenascin W complex</td>\n",
       "      <td>protein_complex</td>\n",
       "      <td>taxon:9606</td>\n",
       "      <td>20170313</td>\n",
       "      <td>ComplexPortal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CPX-1014</td>\n",
       "      <td>GO:0007160</td>\n",
       "      <td>involved_in_XinBP</td>\n",
       "      <td>ComplexPortal</td>\n",
       "      <td>tenascin-x_human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO_REF:0000108</td>\n",
       "      <td>IEA</td>\n",
       "      <td>GO:0098639</td>\n",
       "      <td>P</td>\n",
       "      <td>Tenascin-X complex</td>\n",
       "      <td>TNXB trimer|Tenascin X complex|TNX complex|TN-...</td>\n",
       "      <td>protein_complex</td>\n",
       "      <td>taxon:9606</td>\n",
       "      <td>20190504</td>\n",
       "      <td>GOC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   start_id      end_id               type             db  db_object_symbol  \\\n",
       "0  CPX-1012  GO:0030155  involved_in_XinBP  ComplexPortal  tenascin-w_human   \n",
       "1  CPX-1012  GO:0030334  involved_in_XinBP  ComplexPortal  tenascin-w_human   \n",
       "2  CPX-1012  GO:0062023      part_of_XpoCC  ComplexPortal  tenascin-w_human   \n",
       "3  CPX-1012  GO:1903672  involved_in_XinBP  ComplexPortal  tenascin-w_human   \n",
       "4  CPX-1014  GO:0007160  involved_in_XinBP  ComplexPortal  tenascin-x_human   \n",
       "\n",
       "  qualifier    db_reference evidence_code with_or_from aspect  \\\n",
       "0       NaN   PMID:17909022           IDA          NaN      P   \n",
       "1       NaN   PMID:17909022           IDA          NaN      P   \n",
       "2       NaN   PMID:19884327           IDA          NaN      C   \n",
       "3       NaN   PMID:19884327           IDA          NaN      P   \n",
       "4       NaN  GO_REF:0000108           IEA   GO:0098639      P   \n",
       "\n",
       "       db_object_name                                  db_object_synonym  \\\n",
       "0  Tenascin-W complex                     TNN hexamer|Tenascin W complex   \n",
       "1  Tenascin-W complex                     TNN hexamer|Tenascin W complex   \n",
       "2  Tenascin-W complex                     TNN hexamer|Tenascin W complex   \n",
       "3  Tenascin-W complex                     TNN hexamer|Tenascin W complex   \n",
       "4  Tenascin-X complex  TNXB trimer|Tenascin X complex|TNX complex|TN-...   \n",
       "\n",
       "    db_object_type       taxon      date    assigned_by annotation_extension  \\\n",
       "0  protein_complex  taxon:9606  20170313  ComplexPortal                  NaN   \n",
       "1  protein_complex  taxon:9606  20170313  ComplexPortal                  NaN   \n",
       "2  protein_complex  taxon:9606  20170313  ComplexPortal                  NaN   \n",
       "3  protein_complex  taxon:9606  20170313  ComplexPortal                  NaN   \n",
       "4  protein_complex  taxon:9606  20190504            GOC                  NaN   \n",
       "\n",
       "  gene_product_form_id  \n",
       "0                  NaN  \n",
       "1                  NaN  \n",
       "2                  NaN  \n",
       "3                  NaN  \n",
       "4                  NaN  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type_conversion = {k: v.replace('G', 'X') for k, v in type_conversion.items()}\n",
    "\n",
    "edges_go_cpx = go_complex.rename(columns={'db_object_id': 'start_id', 'go_id': 'end_id'})\n",
    "edges_go_cpx['type'] = edges_go_cpx['aspect'].map(type_conversion)\n",
    "edges_go_cpx = edges_go_cpx[fix_col_order_edge(edges_go_cpx)]\n",
    "edges_go_cpx.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "edges_go_cpx = edges_go_cpx[~edges_go_cpx['qualifier'].str.contains('NOT').fillna(False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add all mapped edges\n",
    "new_edges.append(edges_go_cpx.dropna(subset=['start_id', 'end_id']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Need more to anchor complexes into the network... Gene to Complex, and Complex Identifiers..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>complex_ac</th>\n",
       "      <th>recommended_name</th>\n",
       "      <th>aliases_for_complex</th>\n",
       "      <th>taxonomy_identifier</th>\n",
       "      <th>identifiers_and_stoichiometry_of_molecules_in_complex</th>\n",
       "      <th>confidence</th>\n",
       "      <th>experimental_evidence</th>\n",
       "      <th>go_annotations</th>\n",
       "      <th>cross_references</th>\n",
       "      <th>description</th>\n",
       "      <th>complex_properties</th>\n",
       "      <th>complex_assembly</th>\n",
       "      <th>ligand</th>\n",
       "      <th>disease</th>\n",
       "      <th>agonist</th>\n",
       "      <th>antagonist</th>\n",
       "      <th>comment</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CPX-1196</td>\n",
       "      <td>Polybromo-associated SWI/SNF ATP-dependent chr...</td>\n",
       "      <td>Polybromo-associated SWI/SNF ATP-dependent chr...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Q86U86(0)|Q68CP9(0)|Q8WUB8(0)|P60709(0)|O94805...</td>\n",
       "      <td>ECO:0005547(biological system reconstruction e...</td>\n",
       "      <td>-</td>\n",
       "      <td>GO:0016363(nuclear matrix)|GO:2000045(regulati...</td>\n",
       "      <td>pubmed:11790558(see-also)|pubmed:18809673(see-...</td>\n",
       "      <td>An ATP-dependent chromatin remodeling complex ...</td>\n",
       "      <td>MW approximately 2 MDa. May contain 2 instance...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Coffin-Siris syndrome (CSS) [Orphanet:1465]: a...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>psi-mi:\"MI:0469\"(IntAct)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CPX-1201</td>\n",
       "      <td>Neural progenitor-specific SWI/SNF ATP-depende...</td>\n",
       "      <td>neural progenitor-specific BAF ATP-dependent c...</td>\n",
       "      <td>9606</td>\n",
       "      <td>P51531(0)|O14497(0)|Q969G3(0)|Q6STE5(0)|Q8WUB8...</td>\n",
       "      <td>ECO:0005547(biological system reconstruction e...</td>\n",
       "      <td>-</td>\n",
       "      <td>GO:2000045(regulation of G1/S transition of mi...</td>\n",
       "      <td>pubmed:11790558(see-also)|pubmed:18809673(see-...</td>\n",
       "      <td>An ATP-dependent chromatin remodeling complex ...</td>\n",
       "      <td>MW approximately 2 MDa. May contain 2 instance...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Coffin-Siris syndrome (CSS) [Orphanet:1465]: a...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>psi-mi:\"MI:0469\"(IntAct)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  complex_ac                                   recommended_name  \\\n",
       "0   CPX-1196  Polybromo-associated SWI/SNF ATP-dependent chr...   \n",
       "1   CPX-1201  Neural progenitor-specific SWI/SNF ATP-depende...   \n",
       "\n",
       "                                 aliases_for_complex  taxonomy_identifier  \\\n",
       "0  Polybromo-associated SWI/SNF ATP-dependent chr...                 9606   \n",
       "1  neural progenitor-specific BAF ATP-dependent c...                 9606   \n",
       "\n",
       "  identifiers_and_stoichiometry_of_molecules_in_complex  \\\n",
       "0  Q86U86(0)|Q68CP9(0)|Q8WUB8(0)|P60709(0)|O94805...      \n",
       "1  P51531(0)|O14497(0)|Q969G3(0)|Q6STE5(0)|Q8WUB8...      \n",
       "\n",
       "                                          confidence experimental_evidence  \\\n",
       "0  ECO:0005547(biological system reconstruction e...                     -   \n",
       "1  ECO:0005547(biological system reconstruction e...                     -   \n",
       "\n",
       "                                      go_annotations  \\\n",
       "0  GO:0016363(nuclear matrix)|GO:2000045(regulati...   \n",
       "1  GO:2000045(regulation of G1/S transition of mi...   \n",
       "\n",
       "                                    cross_references  \\\n",
       "0  pubmed:11790558(see-also)|pubmed:18809673(see-...   \n",
       "1  pubmed:11790558(see-also)|pubmed:18809673(see-...   \n",
       "\n",
       "                                         description  \\\n",
       "0  An ATP-dependent chromatin remodeling complex ...   \n",
       "1  An ATP-dependent chromatin remodeling complex ...   \n",
       "\n",
       "                                  complex_properties complex_assembly ligand  \\\n",
       "0  MW approximately 2 MDa. May contain 2 instance...                -      -   \n",
       "1  MW approximately 2 MDa. May contain 2 instance...                -      -   \n",
       "\n",
       "                                             disease agonist antagonist  \\\n",
       "0  Coffin-Siris syndrome (CSS) [Orphanet:1465]: a...       -          -   \n",
       "1  Coffin-Siris syndrome (CSS) [Orphanet:1465]: a...       -          -   \n",
       "\n",
       "  comment                    source  \n",
       "0       -  psi-mi:\"MI:0469\"(IntAct)  \n",
       "1       -  psi-mi:\"MI:0469\"(IntAct)  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cplx = pd.read_csv(load_dir.joinpath('homo_sapiens.tsv'), sep='\\t')\n",
    "cplx.columns = regularize_colnames(cplx.columns)\n",
    "cplx.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CPX-1196</td>\n",
       "      <td>Polybromo-associated SWI/SNF ATP-dependent chr...</td>\n",
       "      <td>Protein Complex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CPX-1201</td>\n",
       "      <td>Neural progenitor-specific SWI/SNF ATP-depende...</td>\n",
       "      <td>Protein Complex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CPX-1194</td>\n",
       "      <td>Muscle cell-specific SWI/SNF ATP-dependent chr...</td>\n",
       "      <td>Protein Complex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CPX-1282</td>\n",
       "      <td>Laminin211-nidogen complex</td>\n",
       "      <td>Protein Complex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CPX-1285</td>\n",
       "      <td>Laminin221-nidogen complex</td>\n",
       "      <td>Protein Complex</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         id                                               name  \\\n",
       "0  CPX-1196  Polybromo-associated SWI/SNF ATP-dependent chr...   \n",
       "1  CPX-1201  Neural progenitor-specific SWI/SNF ATP-depende...   \n",
       "2  CPX-1194  Muscle cell-specific SWI/SNF ATP-dependent chr...   \n",
       "3  CPX-1282                        Laminin211-nidogen complex    \n",
       "4  CPX-1285                        Laminin221-nidogen complex    \n",
       "\n",
       "             label  \n",
       "0  Protein Complex  \n",
       "1  Protein Complex  \n",
       "2  Protein Complex  \n",
       "3  Protein Complex  \n",
       "4  Protein Complex  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cpx_nodes = cplx.rename(columns={'complex_ac': 'id', 'recommended_name': 'name'})\n",
    "cpx_nodes['label'] = 'Protein Complex'\n",
    "cpx_nodes[['id', 'name', 'label']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_nodes.append(cpx_nodes[['id', 'name', 'label']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>complex_ac</th>\n",
       "      <th>recommended_name</th>\n",
       "      <th>aliases_for_complex</th>\n",
       "      <th>taxonomy_identifier</th>\n",
       "      <th>identifiers_and_stoichiometry_of_molecules_in_complex</th>\n",
       "      <th>confidence</th>\n",
       "      <th>experimental_evidence</th>\n",
       "      <th>go_annotations</th>\n",
       "      <th>cross_references</th>\n",
       "      <th>description</th>\n",
       "      <th>complex_properties</th>\n",
       "      <th>complex_assembly</th>\n",
       "      <th>ligand</th>\n",
       "      <th>disease</th>\n",
       "      <th>agonist</th>\n",
       "      <th>antagonist</th>\n",
       "      <th>comment</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CPX-1196</td>\n",
       "      <td>Polybromo-associated SWI/SNF ATP-dependent chr...</td>\n",
       "      <td>Polybromo-associated SWI/SNF ATP-dependent chr...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Q86U86(0)</td>\n",
       "      <td>ECO:0005547(biological system reconstruction e...</td>\n",
       "      <td>-</td>\n",
       "      <td>GO:0016363(nuclear matrix)|GO:2000045(regulati...</td>\n",
       "      <td>pubmed:11790558(see-also)|pubmed:18809673(see-...</td>\n",
       "      <td>An ATP-dependent chromatin remodeling complex ...</td>\n",
       "      <td>MW approximately 2 MDa. May contain 2 instance...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Coffin-Siris syndrome (CSS) [Orphanet:1465]: a...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>psi-mi:\"MI:0469\"(IntAct)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CPX-1196</td>\n",
       "      <td>Polybromo-associated SWI/SNF ATP-dependent chr...</td>\n",
       "      <td>Polybromo-associated SWI/SNF ATP-dependent chr...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Q68CP9(0)</td>\n",
       "      <td>ECO:0005547(biological system reconstruction e...</td>\n",
       "      <td>-</td>\n",
       "      <td>GO:0016363(nuclear matrix)|GO:2000045(regulati...</td>\n",
       "      <td>pubmed:11790558(see-also)|pubmed:18809673(see-...</td>\n",
       "      <td>An ATP-dependent chromatin remodeling complex ...</td>\n",
       "      <td>MW approximately 2 MDa. May contain 2 instance...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Coffin-Siris syndrome (CSS) [Orphanet:1465]: a...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>psi-mi:\"MI:0469\"(IntAct)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  complex_ac                                   recommended_name  \\\n",
       "0   CPX-1196  Polybromo-associated SWI/SNF ATP-dependent chr...   \n",
       "1   CPX-1196  Polybromo-associated SWI/SNF ATP-dependent chr...   \n",
       "\n",
       "                                 aliases_for_complex  taxonomy_identifier  \\\n",
       "0  Polybromo-associated SWI/SNF ATP-dependent chr...                 9606   \n",
       "1  Polybromo-associated SWI/SNF ATP-dependent chr...                 9606   \n",
       "\n",
       "  identifiers_and_stoichiometry_of_molecules_in_complex  \\\n",
       "0                                          Q86U86(0)      \n",
       "1                                          Q68CP9(0)      \n",
       "\n",
       "                                          confidence experimental_evidence  \\\n",
       "0  ECO:0005547(biological system reconstruction e...                     -   \n",
       "1  ECO:0005547(biological system reconstruction e...                     -   \n",
       "\n",
       "                                      go_annotations  \\\n",
       "0  GO:0016363(nuclear matrix)|GO:2000045(regulati...   \n",
       "1  GO:0016363(nuclear matrix)|GO:2000045(regulati...   \n",
       "\n",
       "                                    cross_references  \\\n",
       "0  pubmed:11790558(see-also)|pubmed:18809673(see-...   \n",
       "1  pubmed:11790558(see-also)|pubmed:18809673(see-...   \n",
       "\n",
       "                                         description  \\\n",
       "0  An ATP-dependent chromatin remodeling complex ...   \n",
       "1  An ATP-dependent chromatin remodeling complex ...   \n",
       "\n",
       "                                  complex_properties complex_assembly ligand  \\\n",
       "0  MW approximately 2 MDa. May contain 2 instance...                -      -   \n",
       "1  MW approximately 2 MDa. May contain 2 instance...                -      -   \n",
       "\n",
       "                                             disease agonist antagonist  \\\n",
       "0  Coffin-Siris syndrome (CSS) [Orphanet:1465]: a...       -          -   \n",
       "1  Coffin-Siris syndrome (CSS) [Orphanet:1465]: a...       -          -   \n",
       "\n",
       "  comment                    source  \n",
       "0       -  psi-mi:\"MI:0469\"(IntAct)  \n",
       "1       -  psi-mi:\"MI:0469\"(IntAct)  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cpx_exp = expand_col_on_char(cplx, 'identifiers_and_stoichiometry_of_molecules_in_complex', '|')\n",
    "cpx_exp.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "cpx_exp['uniprot_id'] = cpx_exp['identifiers_and_stoichiometry_of_molecules_in_complex'].str.split('(', expand=True)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Uniprot to gene is one to many, so mapping this way should be fine\n",
    "uniprot_to_geneid_a = expand_col_on_char(all_nodes[['id', 'uni_prot_ids']].dropna(), 'uni_prot_ids', '|')\n",
    "uniprot_to_geneid_a = uniprot_to_geneid_a.set_index('uni_prot_ids')['id'].to_dict()\n",
    "\n",
    "uniprot_to_geneid_b = expand_col_on_char(all_nodes[['id', 'uniprot_id']].dropna(), 'uniprot_id', '|')\n",
    "uniprot_to_geneid_b = uniprot_to_geneid_b.set_index('uniprot_id')['id'].to_dict()\n",
    "\n",
    "uniprot_to_geneid = {**uniprot_to_geneid_a, **uniprot_to_geneid_b}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_id</th>\n",
       "      <th>end_id</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>55193</td>\n",
       "      <td>CPX-1196</td>\n",
       "      <td>part_of_GpoX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>196528</td>\n",
       "      <td>CPX-1196</td>\n",
       "      <td>part_of_GpoX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>55274</td>\n",
       "      <td>CPX-1196</td>\n",
       "      <td>part_of_GpoX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>60</td>\n",
       "      <td>CPX-1196</td>\n",
       "      <td>part_of_GpoX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>51412</td>\n",
       "      <td>CPX-1196</td>\n",
       "      <td>part_of_GpoX</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  start_id    end_id          type\n",
       "0    55193  CPX-1196  part_of_GpoX\n",
       "1   196528  CPX-1196  part_of_GpoX\n",
       "2    55274  CPX-1196  part_of_GpoX\n",
       "3       60  CPX-1196  part_of_GpoX\n",
       "4    51412  CPX-1196  part_of_GpoX"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges_cpx_gene = cpx_exp[['complex_ac', 'uniprot_id']].rename(columns={'complex_ac': 'end_id'})\n",
    "edges_cpx_gene['start_id'] = edges_cpx_gene['uniprot_id'].map(uniprot_to_geneid)\n",
    "edges_cpx_gene['type'] = 'part_of_GpoX'\n",
    "edges_cpx_gene[['start_id', 'end_id', 'type']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_edges.append(edges_cpx_gene[['start_id', 'end_id', 'type']].dropna())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### RNA to GO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>db</th>\n",
       "      <th>db_object_id</th>\n",
       "      <th>db_object_symbol</th>\n",
       "      <th>qualifier</th>\n",
       "      <th>go_id</th>\n",
       "      <th>db_reference</th>\n",
       "      <th>evidence_code</th>\n",
       "      <th>with_or_from</th>\n",
       "      <th>aspect</th>\n",
       "      <th>db_object_name</th>\n",
       "      <th>db_object_synonym</th>\n",
       "      <th>db_object_type</th>\n",
       "      <th>taxon</th>\n",
       "      <th>date</th>\n",
       "      <th>assigned_by</th>\n",
       "      <th>annotation_extension</th>\n",
       "      <th>gene_product_form_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>URS0000001346_9606</td>\n",
       "      <td>URS0000001346_9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0006412</td>\n",
       "      <td>GO_REF:0000108</td>\n",
       "      <td>IEA</td>\n",
       "      <td>GO:0030533</td>\n",
       "      <td>P</td>\n",
       "      <td>Homo sapiens (human) tRNA-Lys</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>taxon:9606</td>\n",
       "      <td>20190504</td>\n",
       "      <td>GOC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>URS0000001346_9606</td>\n",
       "      <td>URS0000001346_9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0030533</td>\n",
       "      <td>GO_REF:0000115</td>\n",
       "      <td>IEA</td>\n",
       "      <td>Rfam:RF00005</td>\n",
       "      <td>F</td>\n",
       "      <td>Homo sapiens (human) tRNA-Lys</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>taxon:9606</td>\n",
       "      <td>20190504</td>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           db        db_object_id    db_object_symbol qualifier       go_id  \\\n",
       "0  RNAcentral  URS0000001346_9606  URS0000001346_9606       NaN  GO:0006412   \n",
       "1  RNAcentral  URS0000001346_9606  URS0000001346_9606       NaN  GO:0030533   \n",
       "\n",
       "     db_reference evidence_code  with_or_from aspect  \\\n",
       "0  GO_REF:0000108           IEA    GO:0030533      P   \n",
       "1  GO_REF:0000115           IEA  Rfam:RF00005      F   \n",
       "\n",
       "                  db_object_name db_object_synonym db_object_type       taxon  \\\n",
       "0  Homo sapiens (human) tRNA-Lys               NaN           tRNA  taxon:9606   \n",
       "1  Homo sapiens (human) tRNA-Lys               NaN           tRNA  taxon:9606   \n",
       "\n",
       "       date assigned_by annotation_extension gene_product_form_id  \n",
       "0  20190504         GOC                  NaN                  NaN  \n",
       "1  20190504  RNAcentral                  NaN                  NaN  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_rna.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_id</th>\n",
       "      <th>end_id</th>\n",
       "      <th>type</th>\n",
       "      <th>db</th>\n",
       "      <th>db_object_symbol</th>\n",
       "      <th>qualifier</th>\n",
       "      <th>db_reference</th>\n",
       "      <th>evidence_code</th>\n",
       "      <th>with_or_from</th>\n",
       "      <th>aspect</th>\n",
       "      <th>db_object_name</th>\n",
       "      <th>db_object_synonym</th>\n",
       "      <th>db_object_type</th>\n",
       "      <th>taxon</th>\n",
       "      <th>date</th>\n",
       "      <th>assigned_by</th>\n",
       "      <th>annotation_extension</th>\n",
       "      <th>gene_product_form_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>URS0000001346_9606</td>\n",
       "      <td>GO:0006412</td>\n",
       "      <td>involved_in_NinBP</td>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>URS0000001346_9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO_REF:0000108</td>\n",
       "      <td>IEA</td>\n",
       "      <td>GO:0030533</td>\n",
       "      <td>P</td>\n",
       "      <td>Homo sapiens (human) tRNA-Lys</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>taxon:9606</td>\n",
       "      <td>20190504</td>\n",
       "      <td>GOC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>URS0000001346_9606</td>\n",
       "      <td>GO:0030533</td>\n",
       "      <td>enables_NeMF</td>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>URS0000001346_9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO_REF:0000115</td>\n",
       "      <td>IEA</td>\n",
       "      <td>Rfam:RF00005</td>\n",
       "      <td>F</td>\n",
       "      <td>Homo sapiens (human) tRNA-Lys</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>taxon:9606</td>\n",
       "      <td>20190504</td>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>URS000000192A_9606</td>\n",
       "      <td>GO:0035068</td>\n",
       "      <td>part_of_NpoCC</td>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>URS000000192A_9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO_REF:0000115</td>\n",
       "      <td>IEA</td>\n",
       "      <td>Rfam:RF00951</td>\n",
       "      <td>C</td>\n",
       "      <td>Homo sapiens (human) MIR1302-2 host gene (MIR1...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lnc_RNA</td>\n",
       "      <td>taxon:9606</td>\n",
       "      <td>20190504</td>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>URS000000192A_9606</td>\n",
       "      <td>GO:0035195</td>\n",
       "      <td>involved_in_NinBP</td>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>URS000000192A_9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO_REF:0000115</td>\n",
       "      <td>IEA</td>\n",
       "      <td>Rfam:RF00951</td>\n",
       "      <td>P</td>\n",
       "      <td>Homo sapiens (human) MIR1302-2 host gene (MIR1...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lnc_RNA</td>\n",
       "      <td>taxon:9606</td>\n",
       "      <td>20190504</td>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>URS00000019BC_9606</td>\n",
       "      <td>GO:0000244</td>\n",
       "      <td>involved_in_NinBP</td>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>URS00000019BC_9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO_REF:0000115</td>\n",
       "      <td>IEA</td>\n",
       "      <td>Rfam:RF00026</td>\n",
       "      <td>P</td>\n",
       "      <td>Homo sapiens (human) snRNA-U6-related</td>\n",
       "      <td>NaN</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>taxon:9606</td>\n",
       "      <td>20190504</td>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             start_id      end_id               type          db  \\\n",
       "0  URS0000001346_9606  GO:0006412  involved_in_NinBP  RNAcentral   \n",
       "1  URS0000001346_9606  GO:0030533       enables_NeMF  RNAcentral   \n",
       "2  URS000000192A_9606  GO:0035068      part_of_NpoCC  RNAcentral   \n",
       "3  URS000000192A_9606  GO:0035195  involved_in_NinBP  RNAcentral   \n",
       "4  URS00000019BC_9606  GO:0000244  involved_in_NinBP  RNAcentral   \n",
       "\n",
       "     db_object_symbol qualifier    db_reference evidence_code  with_or_from  \\\n",
       "0  URS0000001346_9606       NaN  GO_REF:0000108           IEA    GO:0030533   \n",
       "1  URS0000001346_9606       NaN  GO_REF:0000115           IEA  Rfam:RF00005   \n",
       "2  URS000000192A_9606       NaN  GO_REF:0000115           IEA  Rfam:RF00951   \n",
       "3  URS000000192A_9606       NaN  GO_REF:0000115           IEA  Rfam:RF00951   \n",
       "4  URS00000019BC_9606       NaN  GO_REF:0000115           IEA  Rfam:RF00026   \n",
       "\n",
       "  aspect                                     db_object_name db_object_synonym  \\\n",
       "0      P                      Homo sapiens (human) tRNA-Lys               NaN   \n",
       "1      F                      Homo sapiens (human) tRNA-Lys               NaN   \n",
       "2      C  Homo sapiens (human) MIR1302-2 host gene (MIR1...               NaN   \n",
       "3      P  Homo sapiens (human) MIR1302-2 host gene (MIR1...               NaN   \n",
       "4      P              Homo sapiens (human) snRNA-U6-related               NaN   \n",
       "\n",
       "  db_object_type       taxon      date assigned_by annotation_extension  \\\n",
       "0           tRNA  taxon:9606  20190504         GOC                  NaN   \n",
       "1           tRNA  taxon:9606  20190504  RNAcentral                  NaN   \n",
       "2        lnc_RNA  taxon:9606  20190504  RNAcentral                  NaN   \n",
       "3        lnc_RNA  taxon:9606  20190504  RNAcentral                  NaN   \n",
       "4          snRNA  taxon:9606  20190504  RNAcentral                  NaN   \n",
       "\n",
       "  gene_product_form_id  \n",
       "0                  NaN  \n",
       "1                  NaN  \n",
       "2                  NaN  \n",
       "3                  NaN  \n",
       "4                  NaN  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type_conversion = {k: v.replace('X', 'N') for k, v in type_conversion.items()}\n",
    "edges_go_rna = go_rna.rename(columns={'db_object_id': 'start_id', 'go_id': 'end_id'})\n",
    "edges_go_rna['type'] = edges_go_rna['aspect'].map(type_conversion)\n",
    "edges_go_rna = edges_go_rna[fix_col_order_edge(edges_go_rna)]\n",
    "edges_go_rna.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "edges_go_rna = edges_go_rna[~edges_go_rna['qualifier'].str.contains('NOT').fillna(False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_edges.append(edges_go_rna)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Need some more info to add RNA as node type, specifically names\n",
    "\n",
    "Some other DBs with interesting MicroRNA related edges use the name as an identifier rather than an RNA Central ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>db</th>\n",
       "      <th>db_object_id</th>\n",
       "      <th>db_object_symbol</th>\n",
       "      <th>db_object_name</th>\n",
       "      <th>db_object_synonyms</th>\n",
       "      <th>db_object_type</th>\n",
       "      <th>taxon</th>\n",
       "      <th>parent_object_id</th>\n",
       "      <th>db_xrefs</th>\n",
       "      <th>properties</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>URS00006753F8_4081</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Solanum lycopersicum (tomato) tRNA-Tyr for ant...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>taxon:4081</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>URS0000675402_7159</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Aedes aegypti tRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>taxon:7159</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>URS0000675413_6945</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ixodes scapularis tRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>taxon:6945</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>URS0000675414_59463</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Myotis lucifugus (little brown bat) snRNA U6 s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>taxon:59463</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>URS000067541A_60711</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Chlorocebus sabaeus Small nucleolar RNA U13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>snoRNA</td>\n",
       "      <td>taxon:60711</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>URS000067541A_9544</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Macaca mulatta Small nucleolar RNA U13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>snoRNA</td>\n",
       "      <td>taxon:9544</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>URS000067541C_4558</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sorghum bicolor Plant small nucleolar RNA R71</td>\n",
       "      <td>NaN</td>\n",
       "      <td>snoRNA</td>\n",
       "      <td>taxon:4558</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>URS000067541F_1696176</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pelagibacteraceae bacterium GOM-A4 bablM sRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>taxon:1696176</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>URS0000675420_15368</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Brachypodium distachyon microRNA MIR1122</td>\n",
       "      <td>NaN</td>\n",
       "      <td>primary_transcript</td>\n",
       "      <td>taxon:15368</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>URS0000675421_9483</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Callithrix jacchus (white-tufted-ear marmoset)...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>taxon:9483</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           db           db_object_id db_object_symbol  \\\n",
       "0  RNAcentral     URS00006753F8_4081              NaN   \n",
       "1  RNAcentral     URS0000675402_7159              NaN   \n",
       "2  RNAcentral     URS0000675413_6945              NaN   \n",
       "3  RNAcentral    URS0000675414_59463              NaN   \n",
       "4  RNAcentral    URS000067541A_60711              NaN   \n",
       "5  RNAcentral     URS000067541A_9544              NaN   \n",
       "6  RNAcentral     URS000067541C_4558              NaN   \n",
       "7  RNAcentral  URS000067541F_1696176              NaN   \n",
       "8  RNAcentral    URS0000675420_15368              NaN   \n",
       "9  RNAcentral     URS0000675421_9483              NaN   \n",
       "\n",
       "                                      db_object_name db_object_synonyms  \\\n",
       "0  Solanum lycopersicum (tomato) tRNA-Tyr for ant...                NaN   \n",
       "1                                 Aedes aegypti tRNA                NaN   \n",
       "2                             Ixodes scapularis tRNA                NaN   \n",
       "3  Myotis lucifugus (little brown bat) snRNA U6 s...                NaN   \n",
       "4        Chlorocebus sabaeus Small nucleolar RNA U13                NaN   \n",
       "5             Macaca mulatta Small nucleolar RNA U13                NaN   \n",
       "6      Sorghum bicolor Plant small nucleolar RNA R71                NaN   \n",
       "7      Pelagibacteraceae bacterium GOM-A4 bablM sRNA                NaN   \n",
       "8           Brachypodium distachyon microRNA MIR1122                NaN   \n",
       "9  Callithrix jacchus (white-tufted-ear marmoset)...                NaN   \n",
       "\n",
       "       db_object_type          taxon parent_object_id db_xrefs properties  \n",
       "0                tRNA     taxon:4081              NaN      NaN        NaN  \n",
       "1                tRNA     taxon:7159              NaN      NaN        NaN  \n",
       "2                tRNA     taxon:6945              NaN      NaN        NaN  \n",
       "3               snRNA    taxon:59463              NaN      NaN        NaN  \n",
       "4              snoRNA    taxon:60711              NaN      NaN        NaN  \n",
       "5              snoRNA     taxon:9544              NaN      NaN        NaN  \n",
       "6              snoRNA     taxon:4558              NaN      NaN        NaN  \n",
       "7               ncRNA  taxon:1696176              NaN      NaN        NaN  \n",
       "8  primary_transcript    taxon:15368              NaN      NaN        NaN  \n",
       "9               snRNA     taxon:9483              NaN      NaN        NaN  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_colnames = ['db', 'db_object_id', 'db_object_symbol', 'db_object_name', 'db_object_synonyms', 'db_object_type',\n",
    " 'taxon', 'parent_object_id', 'db_xrefs', 'properties']\n",
    "\n",
    "rna_c = pd.read_csv(load_dir.joinpath('rnacentral.gpi.gz'), header=None, names=rna_colnames, \n",
    "                    dtype=str, comment='!', sep='\\t')\n",
    "rna_c.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14483979"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(rna_c)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Human microRNA names start with `hsa-`.  However, this DB seems to have a lot of other text in the name than the smile `hsa-` value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "hsa_lines = rna_c['db_object_name'].str.contains('hsa-')\n",
    "rna_c['rna_name'] = rna_c[hsa_lines]['db_object_name'].apply(lambda s: s[s.index('hsa-'):])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>db</th>\n",
       "      <th>db_object_id</th>\n",
       "      <th>db_object_symbol</th>\n",
       "      <th>db_object_name</th>\n",
       "      <th>db_object_synonyms</th>\n",
       "      <th>db_object_type</th>\n",
       "      <th>taxon</th>\n",
       "      <th>parent_object_id</th>\n",
       "      <th>db_xrefs</th>\n",
       "      <th>properties</th>\n",
       "      <th>rna_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>URS00006753F8_4081</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Solanum lycopersicum (tomato) tRNA-Tyr for ant...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>taxon:4081</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RNAcentral</td>\n",
       "      <td>URS0000675402_7159</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Aedes aegypti tRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>taxon:7159</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           db        db_object_id db_object_symbol  \\\n",
       "0  RNAcentral  URS00006753F8_4081              NaN   \n",
       "1  RNAcentral  URS0000675402_7159              NaN   \n",
       "\n",
       "                                      db_object_name db_object_synonyms  \\\n",
       "0  Solanum lycopersicum (tomato) tRNA-Tyr for ant...                NaN   \n",
       "1                                 Aedes aegypti tRNA                NaN   \n",
       "\n",
       "  db_object_type       taxon parent_object_id db_xrefs properties rna_name  \n",
       "0           tRNA  taxon:4081              NaN      NaN        NaN      NaN  \n",
       "1           tRNA  taxon:7159              NaN      NaN        NaN      NaN  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_c.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2548</th>\n",
       "      <td>URS0000676F0F_9606</td>\n",
       "      <td>hsa-mir-891b precursor</td>\n",
       "      <td>Micro RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4431</th>\n",
       "      <td>URS0000678203_9606</td>\n",
       "      <td>hsa-mir-633 precursor</td>\n",
       "      <td>Micro RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8879</th>\n",
       "      <td>URS000067B227_9606</td>\n",
       "      <td>hsa-mir-644a precursor</td>\n",
       "      <td>Micro RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13287</th>\n",
       "      <td>URS000067E0B9_9606</td>\n",
       "      <td>hsa-mir-920 precursor</td>\n",
       "      <td>Micro RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13806</th>\n",
       "      <td>URS000067E604_9606</td>\n",
       "      <td>hsa-mir-422a precursor</td>\n",
       "      <td>Micro RNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       id                    name      label\n",
       "2548   URS0000676F0F_9606  hsa-mir-891b precursor  Micro RNA\n",
       "4431   URS0000678203_9606   hsa-mir-633 precursor  Micro RNA\n",
       "8879   URS000067B227_9606  hsa-mir-644a precursor  Micro RNA\n",
       "13287  URS000067E0B9_9606   hsa-mir-920 precursor  Micro RNA\n",
       "13806  URS000067E604_9606  hsa-mir-422a precursor  Micro RNA"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_nodes = rna_c.dropna(subset=['rna_name'])\n",
    "rna_nodes = rna_nodes.rename(columns={'rna_name': 'name', 'db_object_id': 'id'})\n",
    "rna_nodes['label'] = 'Micro RNA'\n",
    "rna_nodes[['id', 'name', 'label']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_nodes.append(rna_nodes[['id', 'name', 'label']])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Now get some interesting RNA to Gene Endges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_names = rna_c[hsa_lines][~rna_c[hsa_lines]['rna_name'].str.contains('precursor')]['rna_name'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>miRTarBase ID</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>Species (miRNA)</th>\n",
       "      <th>Target Gene</th>\n",
       "      <th>Target Gene (Entrez Gene ID)</th>\n",
       "      <th>Species (Target Gene)</th>\n",
       "      <th>Experiments</th>\n",
       "      <th>Support Type</th>\n",
       "      <th>References (PMID)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MIRT000002</td>\n",
       "      <td>hsa-miR-20a-5p</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>HIF1A</td>\n",
       "      <td>3091</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Luciferase reporter assay//Western blot//North...</td>\n",
       "      <td>Functional MTI</td>\n",
       "      <td>18632605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MIRT000002</td>\n",
       "      <td>hsa-miR-20a-5p</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>HIF1A</td>\n",
       "      <td>3091</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Functional MTI (Weak)</td>\n",
       "      <td>22473208</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  miRTarBase ID           miRNA Species (miRNA) Target Gene  \\\n",
       "0    MIRT000002  hsa-miR-20a-5p    Homo sapiens       HIF1A   \n",
       "1    MIRT000002  hsa-miR-20a-5p    Homo sapiens       HIF1A   \n",
       "\n",
       "   Target Gene (Entrez Gene ID) Species (Target Gene)  \\\n",
       "0                          3091          Homo sapiens   \n",
       "1                          3091          Homo sapiens   \n",
       "\n",
       "                                         Experiments           Support Type  \\\n",
       "0  Luciferase reporter assay//Western blot//North...         Functional MTI   \n",
       "1                                          HITS-CLIP  Functional MTI (Weak)   \n",
       "\n",
       "   References (PMID)  \n",
       "0           18632605  \n",
       "1           22473208  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mti = pd.read_excel(load_dir.joinpath('hsa_MTI.xlsx'))\n",
    "mti.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mirtarbase_id</th>\n",
       "      <th>mirna</th>\n",
       "      <th>species_mirna</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>target_gene_entrez_gene_id</th>\n",
       "      <th>species_target_gene</th>\n",
       "      <th>experiments</th>\n",
       "      <th>support_type</th>\n",
       "      <th>references_pmid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MIRT000002</td>\n",
       "      <td>hsa-miR-20a-5p</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>HIF1A</td>\n",
       "      <td>3091</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Luciferase reporter assay//Western blot//North...</td>\n",
       "      <td>Functional MTI</td>\n",
       "      <td>18632605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MIRT000002</td>\n",
       "      <td>hsa-miR-20a-5p</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>HIF1A</td>\n",
       "      <td>3091</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Functional MTI (Weak)</td>\n",
       "      <td>22473208</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  mirtarbase_id           mirna species_mirna target_gene  \\\n",
       "0    MIRT000002  hsa-miR-20a-5p  Homo sapiens       HIF1A   \n",
       "1    MIRT000002  hsa-miR-20a-5p  Homo sapiens       HIF1A   \n",
       "\n",
       "   target_gene_entrez_gene_id species_target_gene  \\\n",
       "0                        3091        Homo sapiens   \n",
       "1                        3091        Homo sapiens   \n",
       "\n",
       "                                         experiments           support_type  \\\n",
       "0  Luciferase reporter assay//Western blot//North...         Functional MTI   \n",
       "1                                          HITS-CLIP  Functional MTI (Weak)   \n",
       "\n",
       "   references_pmid  \n",
       "0         18632605  \n",
       "1         22473208  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mti.columns = regularize_colnames(mti.columns)\n",
    "mti.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MicroRNA Gene Annotations mappable to RNACentral\n",
      "  Total:      502,652\n",
      "  Mapped:     493,614\n",
      "  Unmapped:     9,125\n"
     ]
    }
   ],
   "source": [
    "gene_ids = all_nodes.query('label == \"Gene\"')['id']\n",
    "print('MicroRNA Gene Annotations mappable to RNACentral')\n",
    "print('  Total:    {:9,}'.format(len(mti)))\n",
    "print('  Mapped:   {:9,}'.format(len(mti.query('mirna in @rna_names'))))\n",
    "print('  Unmapped: {:9,}'.format(\n",
    "    len(mti) - len(mti.query('mirna in @rna_names and target_gene_entrez_gene_id in @gene_ids'))))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_map = rna_nodes.set_index('name')['id'].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_id</th>\n",
       "      <th>end_id</th>\n",
       "      <th>type</th>\n",
       "      <th>mirtarbase_id</th>\n",
       "      <th>mirna</th>\n",
       "      <th>species_mirna</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>species_target_gene</th>\n",
       "      <th>experiments</th>\n",
       "      <th>support_type</th>\n",
       "      <th>references_pmid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>URS0000574A2C_9606</td>\n",
       "      <td>3091</td>\n",
       "      <td>regulates_NrG</td>\n",
       "      <td>MIRT000002</td>\n",
       "      <td>hsa-miR-20a-5p</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>HIF1A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Luciferase reporter assay//Western blot//North...</td>\n",
       "      <td>Functional MTI</td>\n",
       "      <td>18632605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>URS0000574A2C_9606</td>\n",
       "      <td>3091</td>\n",
       "      <td>regulates_NrG</td>\n",
       "      <td>MIRT000002</td>\n",
       "      <td>hsa-miR-20a-5p</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>HIF1A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Functional MTI (Weak)</td>\n",
       "      <td>22473208</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             start_id end_id           type mirtarbase_id           mirna  \\\n",
       "0  URS0000574A2C_9606   3091  regulates_NrG    MIRT000002  hsa-miR-20a-5p   \n",
       "1  URS0000574A2C_9606   3091  regulates_NrG    MIRT000002  hsa-miR-20a-5p   \n",
       "\n",
       "  species_mirna target_gene species_target_gene  \\\n",
       "0  Homo sapiens       HIF1A        Homo sapiens   \n",
       "1  Homo sapiens       HIF1A        Homo sapiens   \n",
       "\n",
       "                                         experiments           support_type  \\\n",
       "0  Luciferase reporter assay//Western blot//North...         Functional MTI   \n",
       "1                                          HITS-CLIP  Functional MTI (Weak)   \n",
       "\n",
       "  references_pmid  \n",
       "0        18632605  \n",
       "1        22473208  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges_rna_gene = mti.query('mirna in @rna_names and target_gene_entrez_gene_id in @gene_ids').reset_index(drop=True)\n",
    "edges_rna_gene = edges_rna_gene.rename(columns={'target_gene_entrez_gene_id': 'end_id'}) \n",
    "edges_rna_gene['start_id'] = edges_rna_gene['mirna'].map(rna_map)\n",
    "edges_rna_gene['end_id'] = edges_rna_gene['end_id'].astype(str)\n",
    "edges_rna_gene['type'] = 'regulates_NrG'\n",
    "edges_rna_gene = edges_rna_gene[fix_col_order_edge(edges_rna_gene)]\n",
    "edges_rna_gene['references_pmid'] = edges_rna_gene['references_pmid'].astype(str)\n",
    "edges_rna_gene.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_edges.append(edges_rna_gene[['start_id', 'end_id', 'type', 'experiments', 'support_type', 'references_pmid']])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Putting it together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_nodes_df = pd.concat(new_nodes, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CPX-1196</td>\n",
       "      <td>Polybromo-associated SWI/SNF ATP-dependent chr...</td>\n",
       "      <td>Protein Complex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CPX-1201</td>\n",
       "      <td>Neural progenitor-specific SWI/SNF ATP-depende...</td>\n",
       "      <td>Protein Complex</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         id                                               name  \\\n",
       "0  CPX-1196  Polybromo-associated SWI/SNF ATP-dependent chr...   \n",
       "1  CPX-1201  Neural progenitor-specific SWI/SNF ATP-depende...   \n",
       "\n",
       "             label  \n",
       "0  Protein Complex  \n",
       "1  Protein Complex  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_nodes_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_edges_df = pd.concat(new_edges, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "keep_cols = ['start_id', 'end_id', 'type', 'qualifier',\n",
    " 'db_reference', 'evidence_code', 'with_or_from', 'date',\n",
    " 'assigned_by',  'experiments', 'support_type', 'references_pmid']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_id</th>\n",
       "      <th>end_id</th>\n",
       "      <th>type</th>\n",
       "      <th>qualifier</th>\n",
       "      <th>db_reference</th>\n",
       "      <th>evidence_code</th>\n",
       "      <th>with_or_from</th>\n",
       "      <th>date</th>\n",
       "      <th>assigned_by</th>\n",
       "      <th>experiments</th>\n",
       "      <th>support_type</th>\n",
       "      <th>references_pmid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>552891</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>enables_GeMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO_REF:0000002</td>\n",
       "      <td>IEA</td>\n",
       "      <td>InterPro:IPR001770</td>\n",
       "      <td>20190504</td>\n",
       "      <td>InterPro</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>440672</td>\n",
       "      <td>GO:0003723</td>\n",
       "      <td>enables_GeMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO_REF:0000037</td>\n",
       "      <td>IEA</td>\n",
       "      <td>UniProtKB-KW:KW-0694</td>\n",
       "      <td>20190504</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  start_id      end_id          type qualifier    db_reference evidence_code  \\\n",
       "0   552891  GO:0003924  enables_GeMF       NaN  GO_REF:0000002           IEA   \n",
       "1   440672  GO:0003723  enables_GeMF       NaN  GO_REF:0000037           IEA   \n",
       "\n",
       "           with_or_from      date assigned_by experiments support_type  \\\n",
       "0    InterPro:IPR001770  20190504    InterPro         NaN          NaN   \n",
       "1  UniProtKB-KW:KW-0694  20190504     UniProt         NaN          NaN   \n",
       "\n",
       "  references_pmid  \n",
       "0             NaN  \n",
       "1             NaN  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_edges_df[keep_cols].head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of new edges: 1,015,886\n",
      "Number of unique new edges: 697,401\n"
     ]
    }
   ],
   "source": [
    "print('Total number of new edges: {:,}'.format(len(new_edges_df)))\n",
    "print('Number of unique new edges: {:,}'.format(len(new_edges_df.drop_duplicates(subset=['start_id', 'end_id', 'type']))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3min 11s, sys: 733 ms, total: 3min 12s\n",
      "Wall time: 3min 9s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Need to combine those that have multiple of whatever... \n",
    "new_edges_df = combine_group_rows_on_char(new_edges_df, ['start_id', 'end_id', 'type'], ['evidence_code', 'references_pmid'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_edges_df = new_edges_df[keep_cols]\n",
    "new_edges_df['abbv'] = new_edges_df['type'].apply(lambda s: s.split('_')[-1])\n",
    "new_edges_df = new_edges_df.rename(columns={'references_pmid': 'pub_med_ids'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_id</th>\n",
       "      <th>end_id</th>\n",
       "      <th>type</th>\n",
       "      <th>qualifier</th>\n",
       "      <th>db_reference</th>\n",
       "      <th>evidence_code</th>\n",
       "      <th>with_or_from</th>\n",
       "      <th>date</th>\n",
       "      <th>assigned_by</th>\n",
       "      <th>experiments</th>\n",
       "      <th>support_type</th>\n",
       "      <th>pub_med_ids</th>\n",
       "      <th>abbv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GO:0002576</td>\n",
       "      <td>involved_in_GinBP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Reactome:R-HSA-114608</td>\n",
       "      <td>TAS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20181122</td>\n",
       "      <td>Reactome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>GinBP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>GO:0003674</td>\n",
       "      <td>enables_GeMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO_REF:0000015</td>\n",
       "      <td>ND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20070222</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>GeMF</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  start_id      end_id               type qualifier           db_reference  \\\n",
       "0        1  GO:0002576  involved_in_GinBP       NaN  Reactome:R-HSA-114608   \n",
       "1        1  GO:0003674       enables_GeMF       NaN         GO_REF:0000015   \n",
       "\n",
       "  evidence_code with_or_from      date assigned_by experiments support_type  \\\n",
       "0           TAS          NaN  20181122    Reactome         NaN          NaN   \n",
       "1            ND          NaN  20070222     UniProt         NaN          NaN   \n",
       "\n",
       "  pub_med_ids   abbv  \n",
       "0         nan  GinBP  \n",
       "1         nan   GeMF  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_edges_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>label</th>\n",
       "      <th>tree_numbers</th>\n",
       "      <th>drug_bank_ids</th>\n",
       "      <th>alt_disease_ids</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>alt_gene_ids</th>\n",
       "      <th>bio_gridids</th>\n",
       "      <th>pharm_gkbids</th>\n",
       "      <th>uni_prot_ids</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MESH:C089250</td>\n",
       "      <td>(0.017ferrocene)amylose</td>\n",
       "      <td>Compound</td>\n",
       "      <td>D01.490.200/C089250|D02.691.550.200/C089250|D0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MESH:C114385</td>\n",
       "      <td>001-C8-NBD</td>\n",
       "      <td>Compound</td>\n",
       "      <td>D03.383.129.462.580/C114385|D12.644.456/C114385</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             id                     name     label  \\\n",
       "0  MESH:C089250  (0.017ferrocene)amylose  Compound   \n",
       "1  MESH:C114385               001-C8-NBD  Compound   \n",
       "\n",
       "                                        tree_numbers drug_bank_ids  \\\n",
       "0  D01.490.200/C089250|D02.691.550.200/C089250|D0...           NaN   \n",
       "1    D03.383.129.462.580/C114385|D12.644.456/C114385           NaN   \n",
       "\n",
       "  alt_disease_ids gene_symbol alt_gene_ids bio_gridids pharm_gkbids  \\\n",
       "0             NaN         NaN          NaN         NaN          NaN   \n",
       "1             NaN         NaN          NaN         NaN          NaN   \n",
       "\n",
       "  uni_prot_ids uniprot_id  \n",
       "0          NaN        NaN  \n",
       "1          NaN        NaN  "
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_nodes_out = pd.concat([all_nodes, new_nodes_df], sort=False)\n",
    "all_nodes_out.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "753050"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(all_nodes_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_id</th>\n",
       "      <th>end_id</th>\n",
       "      <th>type</th>\n",
       "      <th>parent_ixn</th>\n",
       "      <th>pub_med_ids</th>\n",
       "      <th>organism_id</th>\n",
       "      <th>abbv</th>\n",
       "      <th>direct_evidence</th>\n",
       "      <th>corrected_pvalue</th>\n",
       "      <th>inference_gene_symbol</th>\n",
       "      <th>qualifier</th>\n",
       "      <th>db_reference</th>\n",
       "      <th>evidence_code</th>\n",
       "      <th>with_or_from</th>\n",
       "      <th>date</th>\n",
       "      <th>assigned_by</th>\n",
       "      <th>experiments</th>\n",
       "      <th>support_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MESH:C000121</td>\n",
       "      <td>4313</td>\n",
       "      <td>decreases_activity_CdaG</td>\n",
       "      <td>decreases^activity</td>\n",
       "      <td>25899827</td>\n",
       "      <td>9606</td>\n",
       "      <td>CdaG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MESH:C000121</td>\n",
       "      <td>4313</td>\n",
       "      <td>decreases_expression_CdeG</td>\n",
       "      <td>decreases^expression</td>\n",
       "      <td>25899827</td>\n",
       "      <td>9606</td>\n",
       "      <td>CdeG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       start_id end_id                       type            parent_ixn  \\\n",
       "0  MESH:C000121   4313    decreases_activity_CdaG    decreases^activity   \n",
       "1  MESH:C000121   4313  decreases_expression_CdeG  decreases^expression   \n",
       "\n",
       "  pub_med_ids organism_id  abbv direct_evidence corrected_pvalue  \\\n",
       "0    25899827        9606  CdaG             NaN              NaN   \n",
       "1    25899827        9606  CdeG             NaN              NaN   \n",
       "\n",
       "  inference_gene_symbol qualifier db_reference evidence_code with_or_from  \\\n",
       "0                   NaN       NaN          NaN           NaN          NaN   \n",
       "1                   NaN       NaN          NaN           NaN          NaN   \n",
       "\n",
       "  date assigned_by experiments support_type  \n",
       "0  NaN         NaN         NaN          NaN  \n",
       "1  NaN         NaN         NaN          NaN  "
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_edges_out = pd.concat([edges, new_edges_df], sort=False)\n",
    "all_edges_out.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "node_ids = all_nodes_out['id'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(753050, 753050)"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(all_nodes_out), all_nodes_out['id'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6052217\n",
      "6013920\n"
     ]
    }
   ],
   "source": [
    "# Double check that all of the edges have a startID and End ID contained in the Nodes...\n",
    "print(len(all_edges_out))\n",
    "all_edges_out_filt = all_edges_out.query('start_id in @node_ids and end_id in @node_ids')\n",
    "print(len(all_edges_out_filt))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "part_of_NpoCC        16320\n",
       "involved_in_NinBP    13177\n",
       "enables_NeMF          8799\n",
       "involved_in_GinBP        1\n",
       "Name: type, dtype: int64"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Lests see how many of what kinds of edges had to be dropped...\n",
    "all_edges_out.query('start_id not in @node_ids or end_id not in @node_ids')['type'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These are all RNAs that could not be mapped properly to a name... they also have very minimal information when accessed on RNACentral, and would increase the number of miRNA nodes by a factor of 5."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_edge_ids = all_edges_out_filt[['start_id', 'end_id']].stack().unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "753050\n",
      "97006\n"
     ]
    }
   ],
   "source": [
    "filt_nodes_out = all_nodes_out.query('id in @all_edge_ids')\n",
    "print(len(all_nodes_out))\n",
    "print(len(filt_nodes_out))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Gene                  46777\n",
       "Compound              16347\n",
       "Biological Process    14391\n",
       "Disease                7201\n",
       "Molecular Function     4384\n",
       "Micro RNA              3064\n",
       "Pathway                2363\n",
       "Cellular Component     1835\n",
       "Protein Complex         644\n",
       "Name: label, dtype: int64"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filt_nodes_out['label'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "associated_with_BPawD        1798362\n",
       "associated_with_CawPW        1215923\n",
       "associated_with_DawPW         563169\n",
       "increases_expression_CieG     479825\n",
       "decreases_expression_CdeG     410983\n",
       "regulates_NrG                 373942\n",
       "associated_with_MFawD         199530\n",
       "affects_expression_CaeG       137093\n",
       "involved_in_GinBP             135869\n",
       "part_of_GpoPW                 135809\n",
       "associated_with_CCawD         126497\n",
       "part_of_GpoCC                  78258\n",
       "enables_GeMF                   64281\n",
       "marker_or_mechanism_CmD        62912\n",
       "increases_phenotype_CipBP      35984\n",
       "therapeutic_CtD                34573\n",
       "marker_or_mechanism_GmD        29342\n",
       "increases_activity_CiaG        29001\n",
       "decreases_reaction_CdrBP       26730\n",
       "decreases_activity_CdaG        21926\n",
       "decreases_phenotype_CdpBP      16993\n",
       "affects_phenotype_CapBP        11221\n",
       "increases_reaction_CirBP        6694\n",
       "affects_reaction_CarBP          2419\n",
       "affects_activity_CaaG           2382\n",
       "involved_in_NinBP               2313\n",
       "part_of_GpoX                    2047\n",
       "therapeutic_GtD                 1890\n",
       "decreases_phenotype_CdpMF       1831\n",
       "decreases_reaction_CdrMF        1662\n",
       "increases_phenotype_CipMF       1337\n",
       "involved_in_XinBP                807\n",
       "part_of_NpoCC                    807\n",
       "enables_XeMF                     355\n",
       "enables_NeMF                     270\n",
       "increases_reaction_CirMF         220\n",
       "affects_phenotype_CapMF          189\n",
       "affects_reaction_CarMF           176\n",
       "part_of_XpoCC                    155\n",
       "decreases_phenotype_CdpCC         44\n",
       "increases_phenotype_CipCC         43\n",
       "decreases_reaction_CdrCC          31\n",
       "increases_reaction_CirCC          14\n",
       "affects_phenotype_CapCC            9\n",
       "affects_reaction_CarCC             2\n",
       "Name: type, dtype: int64"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_edges_out_filt['type'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "IPython.notebook.kernel.execute('nb_name = \"' + IPython.notebook.notebook_name + '\"')\n",
       "// hack to get the filename for this notebook"
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%javascript\n",
    "IPython.notebook.kernel.execute('nb_name = \"' + IPython.notebook.notebook_name + '\"')\n",
    "// hack to get the filename for this notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir = Path('../2_pipeline/').joinpath(nb_name.split('.')[0]).joinpath('out').resolve()\n",
    "out_dir.mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "gt.add_colons(new_nodes_df, id_name='identifier').to_csv(out_dir.joinpath('new_nodes.csv'), index=False)\n",
    "gt.add_colons(new_edges_df).to_csv(out_dir.joinpath('new_edges.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "gt.add_colons(all_nodes_out, id_name='identifier').to_csv(out_dir.joinpath('nodes_all.csv'), index=False)\n",
    "gt.add_colons(filt_nodes_out, id_name='identifier').to_csv(out_dir.joinpath('nodes_filt.csv'), index=False)\n",
    "\n",
    "gt.add_colons(all_edges_out_filt).to_csv(out_dir.joinpath('edges.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![Metagraph](../2_pipeline/03_Adding_GO_Annotations/out/CTD_GO_MTIR_metagraph.png)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
