{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GO Nodes into Edges\n",
    "\n",
    "GO Biological Process nodes sometimes have to do with regulation...\n",
    "\n",
    "E.G. GO:0045861\tnegative regulation of proteolysis\t\n",
    "\n",
    "Any Term that links to this go term essentially has an edge:\n",
    "    negatively_regulates -> GO:0006508\tproteolysis\n",
    "\n",
    "Therefore we will determine these regulation nodes and convert them to edges."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "from metapaths.tools.hetnet_file_processing import read_reactome\n",
    "from metapaths.tools.processing import regularize_colnames, head, expand_col_on_char, combine_group_rows_on_char\n",
    "from hetnet_ml.src import graph_tools as gt\n",
    "import tools.obo_tools as ot\n",
    "\n",
    "%matplotlib inline\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "load_dir = Path('../2_pipeline/00_download_data/out/').resolve()\n",
    "network_dir = load_dir.parent.parent.joinpath('05_GO_Genes_and_protein_families/out')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "IPython.notebook.kernel.execute('nb_name = \"' + IPython.notebook.notebook_name + '\"')\n",
       "// hack to get the filename for this notebook"
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%javascript\n",
    "IPython.notebook.kernel.execute('nb_name = \"' + IPython.notebook.notebook_name + '\"')\n",
    "// hack to get the filename for this notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir = Path('../2_pipeline/').joinpath(nb_name.split('.')[0]).joinpath('out').resolve()\n",
    "out_dir.mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_nodes = gt.remove_colons(pd.read_csv(network_dir.joinpath('nodes_all.csv'), dtype=str))\n",
    "edges = gt.remove_colons(pd.read_csv(network_dir.joinpath('edges.csv'), dtype=str))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_node_ids = all_nodes['id'].values\n",
    "new_nodes = []\n",
    "new_edges = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>label</th>\n",
       "      <th>tree_numbers</th>\n",
       "      <th>drug_bank_ids</th>\n",
       "      <th>alt_disease_ids</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>alt_gene_ids</th>\n",
       "      <th>bio_gridids</th>\n",
       "      <th>pharm_gkbids</th>\n",
       "      <th>uni_prot_ids</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>mesh_ids</th>\n",
       "      <th>chebi_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MESH:C089250</td>\n",
       "      <td>(0.017ferrocene)amylose</td>\n",
       "      <td>Compound</td>\n",
       "      <td>D01.490.200/C089250|D02.691.550.200/C089250|D0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MESH:C114385</td>\n",
       "      <td>001-C8-NBD</td>\n",
       "      <td>Compound</td>\n",
       "      <td>D03.383.129.462.580/C114385|D12.644.456/C114385</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             id                     name     label  \\\n",
       "0  MESH:C089250  (0.017ferrocene)amylose  Compound   \n",
       "1  MESH:C114385               001-C8-NBD  Compound   \n",
       "\n",
       "                                        tree_numbers drug_bank_ids  \\\n",
       "0  D01.490.200/C089250|D02.691.550.200/C089250|D0...           NaN   \n",
       "1    D03.383.129.462.580/C114385|D12.644.456/C114385           NaN   \n",
       "\n",
       "  alt_disease_ids gene_symbol alt_gene_ids bio_gridids pharm_gkbids  \\\n",
       "0             NaN         NaN          NaN         NaN          NaN   \n",
       "1             NaN         NaN          NaN         NaN          NaN   \n",
       "\n",
       "  uni_prot_ids uniprot_id mesh_ids chebi_ids  \n",
       "0          NaN        NaN      NaN       NaN  \n",
       "1          NaN        NaN      NaN       NaN  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_nodes.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GO regulation..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "45017"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_terms = all_nodes.query('label in {}'.format(['Molecular Function', 'Cellular Component', 'Biological Process']))\n",
    "len(go_terms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10748"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reg_terms = go_terms['name'].str.contains('regulation of')\n",
    "reg_terms.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "reg_ids = go_terms[reg_terms]['id'].values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## We'll get the regulation Edges from go.obo\n",
    "\n",
    "Need some functions to parse the obo file..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "go_rels = ot.get_ontology_edges(load_dir.joinpath('go.obo'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>src_id</th>\n",
       "      <th>src_name</th>\n",
       "      <th>rel_type</th>\n",
       "      <th>tgt_id</th>\n",
       "      <th>tgt_name</th>\n",
       "      <th>src_src</th>\n",
       "      <th>tgt_src</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0000001</td>\n",
       "      <td>mitochondrion inheritance</td>\n",
       "      <td>is_a</td>\n",
       "      <td>GO:0048308</td>\n",
       "      <td>organelle inheritance</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0000001</td>\n",
       "      <td>mitochondrion inheritance</td>\n",
       "      <td>is_a</td>\n",
       "      <td>GO:0048311</td>\n",
       "      <td>mitochondrion distribution</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0000002</td>\n",
       "      <td>mitochondrial genome maintenance</td>\n",
       "      <td>is_a</td>\n",
       "      <td>GO:0007005</td>\n",
       "      <td>mitochondrion organization</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0000003</td>\n",
       "      <td>reproduction</td>\n",
       "      <td>is_a</td>\n",
       "      <td>GO:0008150</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0000006</td>\n",
       "      <td>high-affinity zinc transmembrane transporter a...</td>\n",
       "      <td>is_a</td>\n",
       "      <td>GO:0005385</td>\n",
       "      <td>zinc ion transmembrane transporter activity</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GO:0000007</td>\n",
       "      <td>low-affinity zinc ion transmembrane transporte...</td>\n",
       "      <td>is_a</td>\n",
       "      <td>GO:0005385</td>\n",
       "      <td>zinc ion transmembrane transporter activity</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GO:0000009</td>\n",
       "      <td>alpha-1,6-mannosyltransferase activity</td>\n",
       "      <td>is_a</td>\n",
       "      <td>GO:0000030</td>\n",
       "      <td>mannosyltransferase activity</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GO:0000010</td>\n",
       "      <td>trans-hexaprenyltranstransferase activity</td>\n",
       "      <td>is_a</td>\n",
       "      <td>GO:0016765</td>\n",
       "      <td>transferase activity, transferring alkyl or ar...</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GO:0000011</td>\n",
       "      <td>vacuole inheritance</td>\n",
       "      <td>is_a</td>\n",
       "      <td>GO:0007033</td>\n",
       "      <td>vacuole organization</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GO:0000011</td>\n",
       "      <td>vacuole inheritance</td>\n",
       "      <td>is_a</td>\n",
       "      <td>GO:0048308</td>\n",
       "      <td>organelle inheritance</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       src_id                                           src_name rel_type  \\\n",
       "0  GO:0000001                          mitochondrion inheritance     is_a   \n",
       "1  GO:0000001                          mitochondrion inheritance     is_a   \n",
       "2  GO:0000002                   mitochondrial genome maintenance     is_a   \n",
       "3  GO:0000003                                       reproduction     is_a   \n",
       "4  GO:0000006  high-affinity zinc transmembrane transporter a...     is_a   \n",
       "5  GO:0000007  low-affinity zinc ion transmembrane transporte...     is_a   \n",
       "6  GO:0000009             alpha-1,6-mannosyltransferase activity     is_a   \n",
       "7  GO:0000010          trans-hexaprenyltranstransferase activity     is_a   \n",
       "8  GO:0000011                                vacuole inheritance     is_a   \n",
       "9  GO:0000011                                vacuole inheritance     is_a   \n",
       "\n",
       "       tgt_id                                           tgt_name src_src  \\\n",
       "0  GO:0048308                              organelle inheritance      GO   \n",
       "1  GO:0048311                         mitochondrion distribution      GO   \n",
       "2  GO:0007005                         mitochondrion organization      GO   \n",
       "3  GO:0008150                                 biological_process      GO   \n",
       "4  GO:0005385        zinc ion transmembrane transporter activity      GO   \n",
       "5  GO:0005385        zinc ion transmembrane transporter activity      GO   \n",
       "6  GO:0000030                       mannosyltransferase activity      GO   \n",
       "7  GO:0016765  transferase activity, transferring alkyl or ar...      GO   \n",
       "8  GO:0007033                               vacuole organization      GO   \n",
       "9  GO:0048308                              organelle inheritance      GO   \n",
       "\n",
       "  tgt_src  \n",
       "0      GO  \n",
       "1      GO  \n",
       "2      GO  \n",
       "3      GO  \n",
       "4      GO  \n",
       "5      GO  \n",
       "6      GO  \n",
       "7      GO  \n",
       "8      GO  \n",
       "9      GO  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_rels.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "is_a                    77058\n",
       "part_of                  7933\n",
       "regulates                3579\n",
       "negatively_regulates     3131\n",
       "positively_regulates     3105\n",
       "has_part                  733\n",
       "occurs_in                 193\n",
       "happens_during              7\n",
       "ends_during                 1\n",
       "Name: rel_type, dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_rels['rel_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>src_id</th>\n",
       "      <th>src_name</th>\n",
       "      <th>rel_type</th>\n",
       "      <th>tgt_id</th>\n",
       "      <th>tgt_name</th>\n",
       "      <th>src_src</th>\n",
       "      <th>tgt_src</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>GO:0000100</td>\n",
       "      <td>S-methylmethionine transmembrane transporter a...</td>\n",
       "      <td>part_of</td>\n",
       "      <td>GO:0015806</td>\n",
       "      <td>S-methylmethionine transport</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48808</th>\n",
       "      <td>GO:0048905</td>\n",
       "      <td>anterior lateral line neuromast mantle cell di...</td>\n",
       "      <td>part_of</td>\n",
       "      <td>GO:0048901</td>\n",
       "      <td>anterior lateral line neuromast development</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22613</th>\n",
       "      <td>GO:0021628</td>\n",
       "      <td>olfactory nerve formation</td>\n",
       "      <td>part_of</td>\n",
       "      <td>GO:0021627</td>\n",
       "      <td>olfactory nerve morphogenesis</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60188</th>\n",
       "      <td>GO:0070743</td>\n",
       "      <td>interleukin-23 complex</td>\n",
       "      <td>part_of</td>\n",
       "      <td>GO:0005615</td>\n",
       "      <td>extracellular space</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77249</th>\n",
       "      <td>GO:1901260</td>\n",
       "      <td>peptidyl-lysine hydroxylation involved in bact...</td>\n",
       "      <td>part_of</td>\n",
       "      <td>GO:0072580</td>\n",
       "      <td>bacterial-type EF-P lysine modification</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           src_id                                           src_name rel_type  \\\n",
       "134    GO:0000100  S-methylmethionine transmembrane transporter a...  part_of   \n",
       "48808  GO:0048905  anterior lateral line neuromast mantle cell di...  part_of   \n",
       "22613  GO:0021628                          olfactory nerve formation  part_of   \n",
       "60188  GO:0070743                             interleukin-23 complex  part_of   \n",
       "77249  GO:1901260  peptidyl-lysine hydroxylation involved in bact...  part_of   \n",
       "\n",
       "           tgt_id                                     tgt_name src_src tgt_src  \n",
       "134    GO:0015806                 S-methylmethionine transport      GO      GO  \n",
       "48808  GO:0048901  anterior lateral line neuromast development      GO      GO  \n",
       "22613  GO:0021627                olfactory nerve morphogenesis      GO      GO  \n",
       "60188  GO:0005615                          extracellular space      GO      GO  \n",
       "77249  GO:0072580      bacterial-type EF-P lysine modification      GO      GO  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_rels.query('rel_type == \"part_of\"').sample(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "reg_rels = ['regulates', 'negatively_regulates', 'positively_regulates']\n",
    "reg_res = go_rels.query('rel_type in @reg_rels')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "reg_ids = reg_res['src_id'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "combo = gt.combine_nodes_and_edges(all_nodes, edges)\n",
    "combo_cols = ['start_id', 'start_name', 'start_label', 'type', 'end_id', 'end_name', 'end_label']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of edges containing Regulation terms that cannot be converted to an edge: 62,099\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_id</th>\n",
       "      <th>start_name</th>\n",
       "      <th>start_label</th>\n",
       "      <th>type</th>\n",
       "      <th>end_id</th>\n",
       "      <th>end_name</th>\n",
       "      <th>end_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4589622</th>\n",
       "      <td>GO:0035810</td>\n",
       "      <td>positive regulation of urine volume</td>\n",
       "      <td>Biological Process</td>\n",
       "      <td>associated_with_BPawD</td>\n",
       "      <td>MESH:D009140</td>\n",
       "      <td>Musculoskeletal Diseases</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4763535</th>\n",
       "      <td>GO:0099509</td>\n",
       "      <td>regulation of presynaptic cytosolic calcium io...</td>\n",
       "      <td>Biological Process</td>\n",
       "      <td>associated_with_BPawD</td>\n",
       "      <td>MESH:C563134</td>\n",
       "      <td>Spinocerebellar Ataxia, X-Linked 1</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4666560</th>\n",
       "      <td>GO:0008217</td>\n",
       "      <td>regulation of blood pressure</td>\n",
       "      <td>Biological Process</td>\n",
       "      <td>associated_with_BPawD</td>\n",
       "      <td>MESH:C536056</td>\n",
       "      <td>Osteopetrosis autosomal dominant type 1</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3967266</th>\n",
       "      <td>GO:0045792</td>\n",
       "      <td>negative regulation of cell size</td>\n",
       "      <td>Biological Process</td>\n",
       "      <td>associated_with_BPawD</td>\n",
       "      <td>MESH:D009358</td>\n",
       "      <td>Congenital, Hereditary, and Neonatal Diseases ...</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4361165</th>\n",
       "      <td>GO:0010460</td>\n",
       "      <td>positive regulation of heart rate</td>\n",
       "      <td>Biological Process</td>\n",
       "      <td>associated_with_BPawD</td>\n",
       "      <td>MESH:D052456</td>\n",
       "      <td>Hypoalphalipoproteinemias</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           start_id                                         start_name  \\\n",
       "4589622  GO:0035810                positive regulation of urine volume   \n",
       "4763535  GO:0099509  regulation of presynaptic cytosolic calcium io...   \n",
       "4666560  GO:0008217                       regulation of blood pressure   \n",
       "3967266  GO:0045792                   negative regulation of cell size   \n",
       "4361165  GO:0010460                  positive regulation of heart rate   \n",
       "\n",
       "                start_label                   type        end_id  \\\n",
       "4589622  Biological Process  associated_with_BPawD  MESH:D009140   \n",
       "4763535  Biological Process  associated_with_BPawD  MESH:C563134   \n",
       "4666560  Biological Process  associated_with_BPawD  MESH:C536056   \n",
       "3967266  Biological Process  associated_with_BPawD  MESH:D009358   \n",
       "4361165  Biological Process  associated_with_BPawD  MESH:D052456   \n",
       "\n",
       "                                                  end_name end_label  \n",
       "4589622                           Musculoskeletal Diseases   Disease  \n",
       "4763535                 Spinocerebellar Ataxia, X-Linked 1   Disease  \n",
       "4666560            Osteopetrosis autosomal dominant type 1   Disease  \n",
       "3967266  Congenital, Hereditary, and Neonatal Diseases ...   Disease  \n",
       "4361165                          Hypoalphalipoproteinemias   Disease  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "no_map_terms = go_terms[reg_terms].query('id not in @reg_ids')['id'].values\n",
    "print('Number of edges containing Regulation terms that cannot be converted to an edge: {:,}'.format(\n",
    "            len(combo.query('start_id in @no_map_terms and end_label == \"Disease\"'))))\n",
    "combo.query('start_id in @no_map_terms and end_label == \"Disease\"')[combo_cols].sample(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of edges containing Regulation terms that can become an edge: 656,949\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_id</th>\n",
       "      <th>start_name</th>\n",
       "      <th>start_label</th>\n",
       "      <th>type</th>\n",
       "      <th>end_id</th>\n",
       "      <th>end_name</th>\n",
       "      <th>end_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4350018</th>\n",
       "      <td>GO:0045722</td>\n",
       "      <td>positive regulation of gluconeogenesis</td>\n",
       "      <td>Biological Process</td>\n",
       "      <td>associated_with_BPawD</td>\n",
       "      <td>MESH:D011628</td>\n",
       "      <td>Puberty, Delayed</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4262322</th>\n",
       "      <td>GO:0060355</td>\n",
       "      <td>positive regulation of cell adhesion molecule ...</td>\n",
       "      <td>Biological Process</td>\n",
       "      <td>associated_with_BPawD</td>\n",
       "      <td>MESH:D001172</td>\n",
       "      <td>Arthritis, Rheumatoid</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4483534</th>\n",
       "      <td>GO:0045778</td>\n",
       "      <td>positive regulation of ossification</td>\n",
       "      <td>Biological Process</td>\n",
       "      <td>associated_with_BPawD</td>\n",
       "      <td>MESH:D013119</td>\n",
       "      <td>Spinal Cord Injuries</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4346262</th>\n",
       "      <td>GO:0010628</td>\n",
       "      <td>positive regulation of gene expression</td>\n",
       "      <td>Biological Process</td>\n",
       "      <td>associated_with_BPawD</td>\n",
       "      <td>MESH:D012174</td>\n",
       "      <td>Retinitis Pigmentosa</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4050183</th>\n",
       "      <td>GO:0071638</td>\n",
       "      <td>negative regulation of monocyte chemotactic pr...</td>\n",
       "      <td>Biological Process</td>\n",
       "      <td>associated_with_BPawD</td>\n",
       "      <td>MESH:D010003</td>\n",
       "      <td>Osteoarthritis</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           start_id                                         start_name  \\\n",
       "4350018  GO:0045722             positive regulation of gluconeogenesis   \n",
       "4262322  GO:0060355  positive regulation of cell adhesion molecule ...   \n",
       "4483534  GO:0045778                positive regulation of ossification   \n",
       "4346262  GO:0010628             positive regulation of gene expression   \n",
       "4050183  GO:0071638  negative regulation of monocyte chemotactic pr...   \n",
       "\n",
       "                start_label                   type        end_id  \\\n",
       "4350018  Biological Process  associated_with_BPawD  MESH:D011628   \n",
       "4262322  Biological Process  associated_with_BPawD  MESH:D001172   \n",
       "4483534  Biological Process  associated_with_BPawD  MESH:D013119   \n",
       "4346262  Biological Process  associated_with_BPawD  MESH:D012174   \n",
       "4050183  Biological Process  associated_with_BPawD  MESH:D010003   \n",
       "\n",
       "                      end_name end_label  \n",
       "4350018       Puberty, Delayed   Disease  \n",
       "4262322  Arthritis, Rheumatoid   Disease  \n",
       "4483534   Spinal Cord Injuries   Disease  \n",
       "4346262   Retinitis Pigmentosa   Disease  \n",
       "4050183         Osteoarthritis   Disease  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "map_terms = go_terms.query('id in @reg_ids')['id'].values\n",
    "print('Number of edges containing Regulation terms that can become an edge: {:,}'.format(\n",
    "    len(combo.query('start_id in @map_terms and end_label == \"Disease\"'))))\n",
    "combo.query('start_id in @map_terms and end_label == \"Disease\"')[combo_cols].sample(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "532791\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_id</th>\n",
       "      <th>start_name</th>\n",
       "      <th>start_label</th>\n",
       "      <th>type</th>\n",
       "      <th>end_id</th>\n",
       "      <th>end_name</th>\n",
       "      <th>end_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3516631</th>\n",
       "      <td>GO:0019221</td>\n",
       "      <td>cytokine-mediated signaling pathway</td>\n",
       "      <td>Biological Process</td>\n",
       "      <td>associated_with_BPawD</td>\n",
       "      <td>MESH:D011018</td>\n",
       "      <td>Pneumonia, Pneumococcal</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4147031</th>\n",
       "      <td>GO:0006809</td>\n",
       "      <td>nitric oxide biosynthetic process</td>\n",
       "      <td>Biological Process</td>\n",
       "      <td>associated_with_BPawD</td>\n",
       "      <td>MESH:D009325</td>\n",
       "      <td>Nausea</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4949394</th>\n",
       "      <td>GO:0036466</td>\n",
       "      <td>synaptic vesicle recycling via endosome</td>\n",
       "      <td>Biological Process</td>\n",
       "      <td>associated_with_BPawD</td>\n",
       "      <td>MESH:D000236</td>\n",
       "      <td>Adenoma</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4814280</th>\n",
       "      <td>GO:0003016</td>\n",
       "      <td>respiratory system process</td>\n",
       "      <td>Biological Process</td>\n",
       "      <td>associated_with_BPawD</td>\n",
       "      <td>MESH:D005512</td>\n",
       "      <td>Food Hypersensitivity</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5011555</th>\n",
       "      <td>GO:0016192</td>\n",
       "      <td>vesicle-mediated transport</td>\n",
       "      <td>Biological Process</td>\n",
       "      <td>associated_with_BPawD</td>\n",
       "      <td>MESH:D002779</td>\n",
       "      <td>Cholestasis</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           start_id                               start_name  \\\n",
       "3516631  GO:0019221      cytokine-mediated signaling pathway   \n",
       "4147031  GO:0006809        nitric oxide biosynthetic process   \n",
       "4949394  GO:0036466  synaptic vesicle recycling via endosome   \n",
       "4814280  GO:0003016               respiratory system process   \n",
       "5011555  GO:0016192               vesicle-mediated transport   \n",
       "\n",
       "                start_label                   type        end_id  \\\n",
       "3516631  Biological Process  associated_with_BPawD  MESH:D011018   \n",
       "4147031  Biological Process  associated_with_BPawD  MESH:D009325   \n",
       "4949394  Biological Process  associated_with_BPawD  MESH:D000236   \n",
       "4814280  Biological Process  associated_with_BPawD  MESH:D005512   \n",
       "5011555  Biological Process  associated_with_BPawD  MESH:D002779   \n",
       "\n",
       "                        end_name end_label  \n",
       "3516631  Pneumonia, Pneumococcal   Disease  \n",
       "4147031                   Nausea   Disease  \n",
       "4949394                  Adenoma   Disease  \n",
       "4814280    Food Hypersensitivity   Disease  \n",
       "5011555              Cholestasis   Disease  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tgt_ids = reg_res['tgt_id'].values\n",
    "map_terms1 = go_terms.query('id in @tgt_ids')['id'].values\n",
    "print(len(combo.query('start_id in @map_terms1 and end_label == \"Disease\"')))\n",
    "combo.query('start_id in @map_terms1 and end_label == \"Disease\"')[combo_cols].sample(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "34\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>src_id</th>\n",
       "      <th>src_name</th>\n",
       "      <th>rel_type</th>\n",
       "      <th>tgt_id</th>\n",
       "      <th>tgt_name</th>\n",
       "      <th>src_src</th>\n",
       "      <th>tgt_src</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>74831</th>\n",
       "      <td>GO:1900545</td>\n",
       "      <td>regulation of phenotypic switching by regulati...</td>\n",
       "      <td>regulates</td>\n",
       "      <td>GO:0100057</td>\n",
       "      <td>regulation of phenotypic switching by transcri...</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74370</th>\n",
       "      <td>GO:1900382</td>\n",
       "      <td>regulation of thiamine biosynthetic process by...</td>\n",
       "      <td>regulates</td>\n",
       "      <td>GO:0100016</td>\n",
       "      <td>regulation of thiamine biosynthetic process by...</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74384</th>\n",
       "      <td>GO:1900389</td>\n",
       "      <td>regulation of glucose import by regulation of ...</td>\n",
       "      <td>regulates</td>\n",
       "      <td>GO:0100018</td>\n",
       "      <td>regulation of glucose import by transcription ...</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54755</th>\n",
       "      <td>GO:0060239</td>\n",
       "      <td>positive regulation of signal transduction inv...</td>\n",
       "      <td>positively_regulates</td>\n",
       "      <td>GO:0032005</td>\n",
       "      <td>signal transduction involved in positive regul...</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84747</th>\n",
       "      <td>GO:1904001</td>\n",
       "      <td>positive regulation of pyrimidine-containing c...</td>\n",
       "      <td>positively_regulates</td>\n",
       "      <td>GO:0100068</td>\n",
       "      <td>positive regulation of pyrimidine-containing c...</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           src_id                                           src_name  \\\n",
       "74831  GO:1900545  regulation of phenotypic switching by regulati...   \n",
       "74370  GO:1900382  regulation of thiamine biosynthetic process by...   \n",
       "74384  GO:1900389  regulation of glucose import by regulation of ...   \n",
       "54755  GO:0060239  positive regulation of signal transduction inv...   \n",
       "84747  GO:1904001  positive regulation of pyrimidine-containing c...   \n",
       "\n",
       "                   rel_type      tgt_id  \\\n",
       "74831             regulates  GO:0100057   \n",
       "74370             regulates  GO:0100016   \n",
       "74384             regulates  GO:0100018   \n",
       "54755  positively_regulates  GO:0032005   \n",
       "84747  positively_regulates  GO:0100068   \n",
       "\n",
       "                                                tgt_name src_src tgt_src  \n",
       "74831  regulation of phenotypic switching by transcri...      GO      GO  \n",
       "74370  regulation of thiamine biosynthetic process by...      GO      GO  \n",
       "74384  regulation of glucose import by transcription ...      GO      GO  \n",
       "54755  signal transduction involved in positive regul...      GO      GO  \n",
       "84747  positive regulation of pyrimidine-containing c...      GO      GO  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Regulations of regulation.... not sure what to do here but just drop\n",
    "print(len(reg_res[reg_res['tgt_name'].str.contains('regulation of')]))\n",
    "reg_res[reg_res['tgt_name'].str.contains('regulation of')].sample(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "166\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>src_id</th>\n",
       "      <th>src_name</th>\n",
       "      <th>rel_type</th>\n",
       "      <th>tgt_id</th>\n",
       "      <th>tgt_name</th>\n",
       "      <th>src_src</th>\n",
       "      <th>tgt_src</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6400</th>\n",
       "      <td>GO:0004860</td>\n",
       "      <td>protein kinase inhibitor activity</td>\n",
       "      <td>negatively_regulates</td>\n",
       "      <td>GO:0004672</td>\n",
       "      <td>protein kinase activity</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10716</th>\n",
       "      <td>GO:0008073</td>\n",
       "      <td>ornithine decarboxylase inhibitor activity</td>\n",
       "      <td>negatively_regulates</td>\n",
       "      <td>GO:0004586</td>\n",
       "      <td>ornithine decarboxylase activity</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53312</th>\n",
       "      <td>GO:0052371</td>\n",
       "      <td>regulation by organism of entry into other org...</td>\n",
       "      <td>regulates</td>\n",
       "      <td>GO:0051828</td>\n",
       "      <td>entry into other organism involved in symbioti...</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64057</th>\n",
       "      <td>GO:0072587</td>\n",
       "      <td>DNA topoisomerase (ATP-hydrolyzing) activator ...</td>\n",
       "      <td>positively_regulates</td>\n",
       "      <td>GO:0003918</td>\n",
       "      <td>DNA topoisomerase type II (ATP-hydrolyzing) ac...</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64951</th>\n",
       "      <td>GO:0075235</td>\n",
       "      <td>modulation of zoospore movement on or near host</td>\n",
       "      <td>regulates</td>\n",
       "      <td>GO:0075234</td>\n",
       "      <td>zoospore movement on or near host</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           src_id                                           src_name  \\\n",
       "6400   GO:0004860                  protein kinase inhibitor activity   \n",
       "10716  GO:0008073         ornithine decarboxylase inhibitor activity   \n",
       "53312  GO:0052371  regulation by organism of entry into other org...   \n",
       "64057  GO:0072587  DNA topoisomerase (ATP-hydrolyzing) activator ...   \n",
       "64951  GO:0075235    modulation of zoospore movement on or near host   \n",
       "\n",
       "                   rel_type      tgt_id  \\\n",
       "6400   negatively_regulates  GO:0004672   \n",
       "10716  negatively_regulates  GO:0004586   \n",
       "53312             regulates  GO:0051828   \n",
       "64057  positively_regulates  GO:0003918   \n",
       "64951             regulates  GO:0075234   \n",
       "\n",
       "                                                tgt_name src_src tgt_src  \n",
       "6400                             protein kinase activity      GO      GO  \n",
       "10716                   ornithine decarboxylase activity      GO      GO  \n",
       "53312  entry into other organism involved in symbioti...      GO      GO  \n",
       "64057  DNA topoisomerase type II (ATP-hydrolyzing) ac...      GO      GO  \n",
       "64951                  zoospore movement on or near host      GO      GO  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Things without 'regulation of' that appearantly regulate...\n",
    "print(len(reg_res[~reg_res['src_name'].str.contains('regulation of')]))\n",
    "reg_res[~reg_res['src_name'].str.contains('regulation of')].sample(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['start_id', 'end_id', 'type', 'parent_ixn', 'pub_med_ids',\n",
       "       'organism_id', 'abbv', 'source', 'evidence', 'direct_evidence',\n",
       "       'corrected_pvalue', 'inference_gene_symbol', 'qualifier',\n",
       "       'db_reference', 'evidence_code', 'with_or_from', 'date', 'assigned_by',\n",
       "       'experiments', 'support_type'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "combo_cols += ['abbv', 'parent_ixn', 'db_reference', 'evidence_code', 'source', 'evidence']\n",
    "mrg = pd.merge(combo[combo_cols], reg_res, how='left', left_on='end_id', right_on='src_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'6,478,252'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'{:,}'.format(len(mrg))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_id</th>\n",
       "      <th>start_name</th>\n",
       "      <th>start_label</th>\n",
       "      <th>type</th>\n",
       "      <th>end_id</th>\n",
       "      <th>end_name</th>\n",
       "      <th>end_label</th>\n",
       "      <th>abbv</th>\n",
       "      <th>parent_ixn</th>\n",
       "      <th>db_reference</th>\n",
       "      <th>evidence_code</th>\n",
       "      <th>source</th>\n",
       "      <th>evidence</th>\n",
       "      <th>src_id</th>\n",
       "      <th>src_name</th>\n",
       "      <th>rel_type</th>\n",
       "      <th>tgt_id</th>\n",
       "      <th>tgt_name</th>\n",
       "      <th>src_src</th>\n",
       "      <th>tgt_src</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5354839</th>\n",
       "      <td>100</td>\n",
       "      <td>adenosine deaminase</td>\n",
       "      <td>Gene</td>\n",
       "      <td>involved_in_GinBP</td>\n",
       "      <td>GO:0002636</td>\n",
       "      <td>positive regulation of germinal center formation</td>\n",
       "      <td>Biological Process</td>\n",
       "      <td>GinBP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO_REF:0000107</td>\n",
       "      <td>IEA</td>\n",
       "      <td>GO</td>\n",
       "      <td>computed</td>\n",
       "      <td>GO:0002636</td>\n",
       "      <td>positive regulation of germinal center formation</td>\n",
       "      <td>positively_regulates</td>\n",
       "      <td>GO:0002467</td>\n",
       "      <td>germinal center formation</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5354840</th>\n",
       "      <td>100</td>\n",
       "      <td>adenosine deaminase</td>\n",
       "      <td>Gene</td>\n",
       "      <td>involved_in_GinBP</td>\n",
       "      <td>GO:0002686</td>\n",
       "      <td>negative regulation of leukocyte migration</td>\n",
       "      <td>Biological Process</td>\n",
       "      <td>GinBP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO_REF:0000107</td>\n",
       "      <td>IEA</td>\n",
       "      <td>GO</td>\n",
       "      <td>computed</td>\n",
       "      <td>GO:0002686</td>\n",
       "      <td>negative regulation of leukocyte migration</td>\n",
       "      <td>negatively_regulates</td>\n",
       "      <td>GO:0050900</td>\n",
       "      <td>leukocyte migration</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        start_id           start_name start_label               type  \\\n",
       "5354839      100  adenosine deaminase        Gene  involved_in_GinBP   \n",
       "5354840      100  adenosine deaminase        Gene  involved_in_GinBP   \n",
       "\n",
       "             end_id                                          end_name  \\\n",
       "5354839  GO:0002636  positive regulation of germinal center formation   \n",
       "5354840  GO:0002686        negative regulation of leukocyte migration   \n",
       "\n",
       "                  end_label   abbv parent_ixn    db_reference evidence_code  \\\n",
       "5354839  Biological Process  GinBP        NaN  GO_REF:0000107           IEA   \n",
       "5354840  Biological Process  GinBP        NaN  GO_REF:0000107           IEA   \n",
       "\n",
       "        source  evidence      src_id  \\\n",
       "5354839     GO  computed  GO:0002636   \n",
       "5354840     GO  computed  GO:0002686   \n",
       "\n",
       "                                                 src_name  \\\n",
       "5354839  positive regulation of germinal center formation   \n",
       "5354840        negative regulation of leukocyte migration   \n",
       "\n",
       "                     rel_type      tgt_id                   tgt_name src_src  \\\n",
       "5354839  positively_regulates  GO:0002467  germinal center formation      GO   \n",
       "5354840  negatively_regulates  GO:0050900        leukocyte migration      GO   \n",
       "\n",
       "        tgt_src  \n",
       "5354839      GO  \n",
       "5354840      GO  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# don't wnat to touch the curated edges... they already have direction to them\n",
    "reg_edges = mrg[mrg['parent_ixn'].isnull()]\n",
    "reg_edges = reg_edges.dropna(subset=['tgt_id'])\n",
    "\n",
    "reg_edges.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "id_to_label = all_nodes.set_index('id')['label'].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "abbv, et = gt.get_abbrev_dict_and_edge_tuples(all_nodes, edges.drop_duplicates(subset=['type']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_id</th>\n",
       "      <th>start_name</th>\n",
       "      <th>start_label</th>\n",
       "      <th>type</th>\n",
       "      <th>abbv</th>\n",
       "      <th>parent_ixn</th>\n",
       "      <th>db_reference</th>\n",
       "      <th>evidence_code</th>\n",
       "      <th>source</th>\n",
       "      <th>evidence</th>\n",
       "      <th>src_id</th>\n",
       "      <th>src_name</th>\n",
       "      <th>rel_type</th>\n",
       "      <th>end_id</th>\n",
       "      <th>end_name</th>\n",
       "      <th>src_src</th>\n",
       "      <th>tgt_src</th>\n",
       "      <th>end_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5354839</th>\n",
       "      <td>100</td>\n",
       "      <td>adenosine deaminase</td>\n",
       "      <td>Gene</td>\n",
       "      <td>positively_regulates_GprBP</td>\n",
       "      <td>GprBP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO_REF:0000107</td>\n",
       "      <td>IEA</td>\n",
       "      <td>GO</td>\n",
       "      <td>computed</td>\n",
       "      <td>GO:0002636</td>\n",
       "      <td>positive regulation of germinal center formation</td>\n",
       "      <td>positively_regulates</td>\n",
       "      <td>GO:0002467</td>\n",
       "      <td>germinal center formation</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "      <td>Biological Process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5354840</th>\n",
       "      <td>100</td>\n",
       "      <td>adenosine deaminase</td>\n",
       "      <td>Gene</td>\n",
       "      <td>negatively_regulates_GnrBP</td>\n",
       "      <td>GnrBP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO_REF:0000107</td>\n",
       "      <td>IEA</td>\n",
       "      <td>GO</td>\n",
       "      <td>computed</td>\n",
       "      <td>GO:0002686</td>\n",
       "      <td>negative regulation of leukocyte migration</td>\n",
       "      <td>negatively_regulates</td>\n",
       "      <td>GO:0050900</td>\n",
       "      <td>leukocyte migration</td>\n",
       "      <td>GO</td>\n",
       "      <td>GO</td>\n",
       "      <td>Biological Process</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        start_id           start_name start_label                        type  \\\n",
       "5354839      100  adenosine deaminase        Gene  positively_regulates_GprBP   \n",
       "5354840      100  adenosine deaminase        Gene  negatively_regulates_GnrBP   \n",
       "\n",
       "          abbv parent_ixn    db_reference evidence_code source  evidence  \\\n",
       "5354839  GprBP        NaN  GO_REF:0000107           IEA     GO  computed   \n",
       "5354840  GnrBP        NaN  GO_REF:0000107           IEA     GO  computed   \n",
       "\n",
       "             src_id                                          src_name  \\\n",
       "5354839  GO:0002636  positive regulation of germinal center formation   \n",
       "5354840  GO:0002686        negative regulation of leukocyte migration   \n",
       "\n",
       "                     rel_type      end_id                   end_name src_src  \\\n",
       "5354839  positively_regulates  GO:0002467  germinal center formation      GO   \n",
       "5354840  negatively_regulates  GO:0050900        leukocyte migration      GO   \n",
       "\n",
       "        tgt_src           end_label  \n",
       "5354839      GO  Biological Process  \n",
       "5354840      GO  Biological Process  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reg_edges = reg_edges.drop(['end_id', 'end_label', 'end_name'], axis=1)\n",
    "reg_edges['end_label'] = reg_edges['tgt_id'].map(id_to_label)\n",
    "reg_edges['abbv'] = reg_edges['start_label'].map(abbv) + \\\n",
    "                        reg_edges['rel_type'].apply(lambda s: ''.join([z[0] for z in s.split('_')])) + \\\n",
    "                        reg_edges['end_label'].map(abbv)\n",
    "reg_edges['type'] = reg_edges['rel_type'] + '_' + reg_edges['abbv']\n",
    "reg_edges = reg_edges.rename(columns={'tgt_id':'end_id', 'tgt_name': 'end_name'})\n",
    "reg_edges.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_id</th>\n",
       "      <th>type</th>\n",
       "      <th>end_id</th>\n",
       "      <th>abbv</th>\n",
       "      <th>parent_ixn</th>\n",
       "      <th>db_reference</th>\n",
       "      <th>evidence_code</th>\n",
       "      <th>source</th>\n",
       "      <th>evidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100</td>\n",
       "      <td>positively_regulates_GprBP</td>\n",
       "      <td>GO:0002467</td>\n",
       "      <td>GprBP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO_REF:0000107</td>\n",
       "      <td>IEA</td>\n",
       "      <td>GO</td>\n",
       "      <td>computed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100</td>\n",
       "      <td>negatively_regulates_GnrBP</td>\n",
       "      <td>GO:0050900</td>\n",
       "      <td>GnrBP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO_REF:0000107</td>\n",
       "      <td>IEA</td>\n",
       "      <td>GO</td>\n",
       "      <td>computed</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  start_id                        type      end_id   abbv parent_ixn  \\\n",
       "0      100  positively_regulates_GprBP  GO:0002467  GprBP        NaN   \n",
       "1      100  negatively_regulates_GnrBP  GO:0050900  GnrBP        NaN   \n",
       "\n",
       "     db_reference evidence_code source  evidence  \n",
       "0  GO_REF:0000107           IEA     GO  computed  \n",
       "1  GO_REF:0000107           IEA     GO  computed  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reg_edges = reg_edges[combo_cols].drop(['start_name', 'start_label', 'end_name', 'end_label'], axis=1).reset_index(drop=True)\n",
    "reg_edges.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Putting it all together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_id</th>\n",
       "      <th>type</th>\n",
       "      <th>end_id</th>\n",
       "      <th>abbv</th>\n",
       "      <th>parent_ixn</th>\n",
       "      <th>db_reference</th>\n",
       "      <th>evidence_code</th>\n",
       "      <th>source</th>\n",
       "      <th>evidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100</td>\n",
       "      <td>positively_regulates_GprBP</td>\n",
       "      <td>GO:0002467</td>\n",
       "      <td>GprBP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO_REF:0000107</td>\n",
       "      <td>IEA</td>\n",
       "      <td>GO</td>\n",
       "      <td>computed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100</td>\n",
       "      <td>negatively_regulates_GnrBP</td>\n",
       "      <td>GO:0050900</td>\n",
       "      <td>GnrBP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO_REF:0000107</td>\n",
       "      <td>IEA</td>\n",
       "      <td>GO</td>\n",
       "      <td>computed</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  start_id                        type      end_id   abbv parent_ixn  \\\n",
       "0      100  positively_regulates_GprBP  GO:0002467  GprBP        NaN   \n",
       "1      100  negatively_regulates_GnrBP  GO:0050900  GnrBP        NaN   \n",
       "\n",
       "     db_reference evidence_code source  evidence  \n",
       "0  GO_REF:0000107           IEA     GO  computed  \n",
       "1  GO_REF:0000107           IEA     GO  computed  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reg_edges.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "positively_regulates_GprBP     15578\n",
       "negatively_regulates_GnrBP     10464\n",
       "regulates_GrBP                 10036\n",
       "positively_regulates_GprMF      2144\n",
       "negatively_regulates_GnrMF      1593\n",
       "regulates_PFrBP                  898\n",
       "regulates_GrMF                   851\n",
       "negatively_regulates_NnrBP       737\n",
       "positively_regulates_NprBP       464\n",
       "negatively_regulates_PFnrBP      267\n",
       "positively_regulates_XprBP       197\n",
       "positively_regulates_PFprBP      196\n",
       "negatively_regulates_PFnrMF      111\n",
       "regulates_XrBP                   109\n",
       "positively_regulates_PFprMF       90\n",
       "negatively_regulates_XnrBP        63\n",
       "regulates_PFrMF                   58\n",
       "negatively_regulates_NnrMF        50\n",
       "regulates_NrBP                    48\n",
       "regulates_PWrBP                   46\n",
       "negatively_regulates_RXnrBP       36\n",
       "positively_regulates_PWprBP       30\n",
       "positively_regulates_XprMF        17\n",
       "positively_regulates_RXprBP       16\n",
       "regulates_RXrBP                   13\n",
       "regulates_XrMF                    12\n",
       "positively_regulates_NprMF        12\n",
       "negatively_regulates_PWnrBP       11\n",
       "negatively_regulates_XnrMF        10\n",
       "regulates_PWrMF                    7\n",
       "positively_regulates_RXprMF        6\n",
       "positively_regulates_PWprMF        3\n",
       "negatively_regulates_RXnrMF        2\n",
       "regulates_RXrMF                    1\n",
       "Name: type, dtype: int64"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reg_edges['type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of new edges: 44,176\n",
      "Number of unique new edges: 43,704\n"
     ]
    }
   ],
   "source": [
    "print('Total number of new edges: {:,}'.format(len(reg_edges)))\n",
    "print('Number of unique new edges: {:,}'.format(len(reg_edges.drop_duplicates(subset=['start_id', 'end_id', 'type']))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Why are so many edges duplicated? what kind are they?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_id</th>\n",
       "      <th>type</th>\n",
       "      <th>end_id</th>\n",
       "      <th>abbv</th>\n",
       "      <th>parent_ixn</th>\n",
       "      <th>db_reference</th>\n",
       "      <th>evidence_code</th>\n",
       "      <th>source</th>\n",
       "      <th>evidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>780</th>\n",
       "      <td>10221</td>\n",
       "      <td>negatively_regulates_GnrMF</td>\n",
       "      <td>GO:0004672</td>\n",
       "      <td>GnrMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:15299019</td>\n",
       "      <td>IMP</td>\n",
       "      <td>GO</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>781</th>\n",
       "      <td>10221</td>\n",
       "      <td>negatively_regulates_GnrMF</td>\n",
       "      <td>GO:0004672</td>\n",
       "      <td>GnrMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:15299019</td>\n",
       "      <td>IMP</td>\n",
       "      <td>GO</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1940</th>\n",
       "      <td>10681</td>\n",
       "      <td>positively_regulates_GprMF</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>GprMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:10521509</td>\n",
       "      <td>IDA</td>\n",
       "      <td>GO</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941</th>\n",
       "      <td>10681</td>\n",
       "      <td>positively_regulates_GprMF</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>GprMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:10521509</td>\n",
       "      <td>IDA</td>\n",
       "      <td>GO</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2322</th>\n",
       "      <td>10928</td>\n",
       "      <td>positively_regulates_GprMF</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>GprMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:7673236</td>\n",
       "      <td>IDA|IBA</td>\n",
       "      <td>GO</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2324</th>\n",
       "      <td>10928</td>\n",
       "      <td>positively_regulates_GprMF</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>GprMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:7673236</td>\n",
       "      <td>IDA|IBA</td>\n",
       "      <td>GO</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2626</th>\n",
       "      <td>11142</td>\n",
       "      <td>negatively_regulates_GnrMF</td>\n",
       "      <td>GO:0004691</td>\n",
       "      <td>GnrMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:21873635</td>\n",
       "      <td>IBA</td>\n",
       "      <td>GO</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2628</th>\n",
       "      <td>11142</td>\n",
       "      <td>negatively_regulates_GnrMF</td>\n",
       "      <td>GO:0004691</td>\n",
       "      <td>GnrMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:21873635</td>\n",
       "      <td>IBA</td>\n",
       "      <td>GO</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2805</th>\n",
       "      <td>1124</td>\n",
       "      <td>positively_regulates_GprMF</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>GprMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO_REF:0000002</td>\n",
       "      <td>IEA</td>\n",
       "      <td>GO</td>\n",
       "      <td>computed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2807</th>\n",
       "      <td>1124</td>\n",
       "      <td>positively_regulates_GprMF</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>GprMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO_REF:0000002</td>\n",
       "      <td>IEA</td>\n",
       "      <td>GO</td>\n",
       "      <td>computed</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     start_id                        type      end_id   abbv parent_ixn  \\\n",
       "780     10221  negatively_regulates_GnrMF  GO:0004672  GnrMF        NaN   \n",
       "781     10221  negatively_regulates_GnrMF  GO:0004672  GnrMF        NaN   \n",
       "1940    10681  positively_regulates_GprMF  GO:0003924  GprMF        NaN   \n",
       "1941    10681  positively_regulates_GprMF  GO:0003924  GprMF        NaN   \n",
       "2322    10928  positively_regulates_GprMF  GO:0003924  GprMF        NaN   \n",
       "2324    10928  positively_regulates_GprMF  GO:0003924  GprMF        NaN   \n",
       "2626    11142  negatively_regulates_GnrMF  GO:0004691  GnrMF        NaN   \n",
       "2628    11142  negatively_regulates_GnrMF  GO:0004691  GnrMF        NaN   \n",
       "2805     1124  positively_regulates_GprMF  GO:0003924  GprMF        NaN   \n",
       "2807     1124  positively_regulates_GprMF  GO:0003924  GprMF        NaN   \n",
       "\n",
       "        db_reference evidence_code source  evidence  \n",
       "780    PMID:15299019           IMP     GO   curated  \n",
       "781    PMID:15299019           IMP     GO   curated  \n",
       "1940   PMID:10521509           IDA     GO   curated  \n",
       "1941   PMID:10521509           IDA     GO   curated  \n",
       "2322    PMID:7673236       IDA|IBA     GO   curated  \n",
       "2324    PMID:7673236       IDA|IBA     GO   curated  \n",
       "2626   PMID:21873635           IBA     GO   curated  \n",
       "2628   PMID:21873635           IBA     GO   curated  \n",
       "2805  GO_REF:0000002           IEA     GO  computed  \n",
       "2807  GO_REF:0000002           IEA     GO  computed  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ix = reg_edges.duplicated(keep=False)\n",
    "reg_edges[ix].sort_values(['start_id', 'end_id', 'type']).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "positively_regulates_GprMF     92\n",
       "negatively_regulates_GnrMF     60\n",
       "positively_regulates_PFprMF    24\n",
       "regulates_GrMF                 10\n",
       "negatively_regulates_PFnrMF    10\n",
       "negatively_regulates_XnrMF      2\n",
       "Name: type, dtype: int64"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reg_edges[ix]['type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_id</th>\n",
       "      <th>type</th>\n",
       "      <th>end_id</th>\n",
       "      <th>abbv</th>\n",
       "      <th>parent_ixn</th>\n",
       "      <th>db_reference</th>\n",
       "      <th>evidence_code</th>\n",
       "      <th>source</th>\n",
       "      <th>evidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>780</th>\n",
       "      <td>10221</td>\n",
       "      <td>negatively_regulates_GnrMF</td>\n",
       "      <td>GO:0004672</td>\n",
       "      <td>GnrMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:15299019</td>\n",
       "      <td>IMP</td>\n",
       "      <td>GO</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>781</th>\n",
       "      <td>10221</td>\n",
       "      <td>negatively_regulates_GnrMF</td>\n",
       "      <td>GO:0004672</td>\n",
       "      <td>GnrMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:15299019</td>\n",
       "      <td>IMP</td>\n",
       "      <td>GO</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2626</th>\n",
       "      <td>11142</td>\n",
       "      <td>negatively_regulates_GnrMF</td>\n",
       "      <td>GO:0004691</td>\n",
       "      <td>GnrMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:21873635</td>\n",
       "      <td>IBA</td>\n",
       "      <td>GO</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2628</th>\n",
       "      <td>11142</td>\n",
       "      <td>negatively_regulates_GnrMF</td>\n",
       "      <td>GO:0004691</td>\n",
       "      <td>GnrMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:21873635</td>\n",
       "      <td>IBA</td>\n",
       "      <td>GO</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4045</th>\n",
       "      <td>132864</td>\n",
       "      <td>negatively_regulates_GnrMF</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>GnrMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO_REF:0000024</td>\n",
       "      <td>ISS</td>\n",
       "      <td>GO</td>\n",
       "      <td>computed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4046</th>\n",
       "      <td>132864</td>\n",
       "      <td>negatively_regulates_GnrMF</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>GnrMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO_REF:0000024</td>\n",
       "      <td>ISS</td>\n",
       "      <td>GO</td>\n",
       "      <td>computed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1940</th>\n",
       "      <td>10681</td>\n",
       "      <td>positively_regulates_GprMF</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>GprMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:10521509</td>\n",
       "      <td>IDA</td>\n",
       "      <td>GO</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941</th>\n",
       "      <td>10681</td>\n",
       "      <td>positively_regulates_GprMF</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>GprMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:10521509</td>\n",
       "      <td>IDA</td>\n",
       "      <td>GO</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2322</th>\n",
       "      <td>10928</td>\n",
       "      <td>positively_regulates_GprMF</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>GprMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:7673236</td>\n",
       "      <td>IDA|IBA</td>\n",
       "      <td>GO</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2324</th>\n",
       "      <td>10928</td>\n",
       "      <td>positively_regulates_GprMF</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>GprMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:7673236</td>\n",
       "      <td>IDA|IBA</td>\n",
       "      <td>GO</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2805</th>\n",
       "      <td>1124</td>\n",
       "      <td>positively_regulates_GprMF</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>GprMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO_REF:0000002</td>\n",
       "      <td>IEA</td>\n",
       "      <td>GO</td>\n",
       "      <td>computed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2807</th>\n",
       "      <td>1124</td>\n",
       "      <td>positively_regulates_GprMF</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>GprMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO_REF:0000002</td>\n",
       "      <td>IEA</td>\n",
       "      <td>GO</td>\n",
       "      <td>computed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3732</th>\n",
       "      <td>124997</td>\n",
       "      <td>regulates_GrMF</td>\n",
       "      <td>GO:0035004</td>\n",
       "      <td>GrMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:26783301</td>\n",
       "      <td>IMP</td>\n",
       "      <td>GO</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3733</th>\n",
       "      <td>124997</td>\n",
       "      <td>regulates_GrMF</td>\n",
       "      <td>GO:0035004</td>\n",
       "      <td>GrMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:26783301</td>\n",
       "      <td>IMP</td>\n",
       "      <td>GO</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3933</th>\n",
       "      <td>129138</td>\n",
       "      <td>regulates_GrMF</td>\n",
       "      <td>GO:0004672</td>\n",
       "      <td>GrMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:21873635</td>\n",
       "      <td>IBA</td>\n",
       "      <td>GO</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3935</th>\n",
       "      <td>129138</td>\n",
       "      <td>regulates_GrMF</td>\n",
       "      <td>GO:0004672</td>\n",
       "      <td>GrMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:21873635</td>\n",
       "      <td>IBA</td>\n",
       "      <td>GO</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6831</th>\n",
       "      <td>1956</td>\n",
       "      <td>regulates_GrMF</td>\n",
       "      <td>GO:0004517</td>\n",
       "      <td>GrMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:12828935</td>\n",
       "      <td>IDA</td>\n",
       "      <td>GO</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6856</th>\n",
       "      <td>1956</td>\n",
       "      <td>regulates_GrMF</td>\n",
       "      <td>GO:0004517</td>\n",
       "      <td>GrMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:12828935</td>\n",
       "      <td>IDA</td>\n",
       "      <td>GO</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41065</th>\n",
       "      <td>CPX-98</td>\n",
       "      <td>negatively_regulates_XnrMF</td>\n",
       "      <td>GO:0008233</td>\n",
       "      <td>XnrMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:20860624</td>\n",
       "      <td>IDA</td>\n",
       "      <td>GO</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41066</th>\n",
       "      <td>CPX-98</td>\n",
       "      <td>negatively_regulates_XnrMF</td>\n",
       "      <td>GO:0008233</td>\n",
       "      <td>XnrMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:20860624</td>\n",
       "      <td>IDA</td>\n",
       "      <td>GO</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42701</th>\n",
       "      <td>InterPro:IPR003913</td>\n",
       "      <td>positively_regulates_PFprMF</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>PFprMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>InterPro</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42703</th>\n",
       "      <td>InterPro:IPR003913</td>\n",
       "      <td>positively_regulates_PFprMF</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>PFprMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>InterPro</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43236</th>\n",
       "      <td>InterPro:IPR026817</td>\n",
       "      <td>positively_regulates_PFprMF</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>PFprMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>InterPro</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43237</th>\n",
       "      <td>InterPro:IPR026817</td>\n",
       "      <td>positively_regulates_PFprMF</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>PFprMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>InterPro</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43254</th>\n",
       "      <td>InterPro:IPR027107</td>\n",
       "      <td>positively_regulates_PFprMF</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>PFprMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>InterPro</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43255</th>\n",
       "      <td>InterPro:IPR027107</td>\n",
       "      <td>positively_regulates_PFprMF</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>PFprMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>InterPro</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42782</th>\n",
       "      <td>InterPro:IPR007648</td>\n",
       "      <td>negatively_regulates_PFnrMF</td>\n",
       "      <td>GO:0016887</td>\n",
       "      <td>PFnrMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>InterPro</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42783</th>\n",
       "      <td>InterPro:IPR007648</td>\n",
       "      <td>negatively_regulates_PFnrMF</td>\n",
       "      <td>GO:0016887</td>\n",
       "      <td>PFnrMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>InterPro</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42806</th>\n",
       "      <td>InterPro:IPR008296</td>\n",
       "      <td>negatively_regulates_PFnrMF</td>\n",
       "      <td>GO:0008233</td>\n",
       "      <td>PFnrMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>InterPro</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42807</th>\n",
       "      <td>InterPro:IPR008296</td>\n",
       "      <td>negatively_regulates_PFnrMF</td>\n",
       "      <td>GO:0008233</td>\n",
       "      <td>PFnrMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>InterPro</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42885</th>\n",
       "      <td>InterPro:IPR010832</td>\n",
       "      <td>negatively_regulates_PFnrMF</td>\n",
       "      <td>GO:0004175</td>\n",
       "      <td>PFnrMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>InterPro</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42886</th>\n",
       "      <td>InterPro:IPR010832</td>\n",
       "      <td>negatively_regulates_PFnrMF</td>\n",
       "      <td>GO:0004175</td>\n",
       "      <td>PFnrMF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>InterPro</td>\n",
       "      <td>curated</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 start_id                         type      end_id    abbv  \\\n",
       "780                 10221   negatively_regulates_GnrMF  GO:0004672   GnrMF   \n",
       "781                 10221   negatively_regulates_GnrMF  GO:0004672   GnrMF   \n",
       "2626                11142   negatively_regulates_GnrMF  GO:0004691   GnrMF   \n",
       "2628                11142   negatively_regulates_GnrMF  GO:0004691   GnrMF   \n",
       "4045               132864   negatively_regulates_GnrMF  GO:0003924   GnrMF   \n",
       "4046               132864   negatively_regulates_GnrMF  GO:0003924   GnrMF   \n",
       "1940                10681   positively_regulates_GprMF  GO:0003924   GprMF   \n",
       "1941                10681   positively_regulates_GprMF  GO:0003924   GprMF   \n",
       "2322                10928   positively_regulates_GprMF  GO:0003924   GprMF   \n",
       "2324                10928   positively_regulates_GprMF  GO:0003924   GprMF   \n",
       "2805                 1124   positively_regulates_GprMF  GO:0003924   GprMF   \n",
       "2807                 1124   positively_regulates_GprMF  GO:0003924   GprMF   \n",
       "3732               124997               regulates_GrMF  GO:0035004    GrMF   \n",
       "3733               124997               regulates_GrMF  GO:0035004    GrMF   \n",
       "3933               129138               regulates_GrMF  GO:0004672    GrMF   \n",
       "3935               129138               regulates_GrMF  GO:0004672    GrMF   \n",
       "6831                 1956               regulates_GrMF  GO:0004517    GrMF   \n",
       "6856                 1956               regulates_GrMF  GO:0004517    GrMF   \n",
       "41065              CPX-98   negatively_regulates_XnrMF  GO:0008233   XnrMF   \n",
       "41066              CPX-98   negatively_regulates_XnrMF  GO:0008233   XnrMF   \n",
       "42701  InterPro:IPR003913  positively_regulates_PFprMF  GO:0003924  PFprMF   \n",
       "42703  InterPro:IPR003913  positively_regulates_PFprMF  GO:0003924  PFprMF   \n",
       "43236  InterPro:IPR026817  positively_regulates_PFprMF  GO:0003924  PFprMF   \n",
       "43237  InterPro:IPR026817  positively_regulates_PFprMF  GO:0003924  PFprMF   \n",
       "43254  InterPro:IPR027107  positively_regulates_PFprMF  GO:0003924  PFprMF   \n",
       "43255  InterPro:IPR027107  positively_regulates_PFprMF  GO:0003924  PFprMF   \n",
       "42782  InterPro:IPR007648  negatively_regulates_PFnrMF  GO:0016887  PFnrMF   \n",
       "42783  InterPro:IPR007648  negatively_regulates_PFnrMF  GO:0016887  PFnrMF   \n",
       "42806  InterPro:IPR008296  negatively_regulates_PFnrMF  GO:0008233  PFnrMF   \n",
       "42807  InterPro:IPR008296  negatively_regulates_PFnrMF  GO:0008233  PFnrMF   \n",
       "42885  InterPro:IPR010832  negatively_regulates_PFnrMF  GO:0004175  PFnrMF   \n",
       "42886  InterPro:IPR010832  negatively_regulates_PFnrMF  GO:0004175  PFnrMF   \n",
       "\n",
       "      parent_ixn    db_reference evidence_code    source  evidence  \n",
       "780          NaN   PMID:15299019           IMP        GO   curated  \n",
       "781          NaN   PMID:15299019           IMP        GO   curated  \n",
       "2626         NaN   PMID:21873635           IBA        GO   curated  \n",
       "2628         NaN   PMID:21873635           IBA        GO   curated  \n",
       "4045         NaN  GO_REF:0000024           ISS        GO  computed  \n",
       "4046         NaN  GO_REF:0000024           ISS        GO  computed  \n",
       "1940         NaN   PMID:10521509           IDA        GO   curated  \n",
       "1941         NaN   PMID:10521509           IDA        GO   curated  \n",
       "2322         NaN    PMID:7673236       IDA|IBA        GO   curated  \n",
       "2324         NaN    PMID:7673236       IDA|IBA        GO   curated  \n",
       "2805         NaN  GO_REF:0000002           IEA        GO  computed  \n",
       "2807         NaN  GO_REF:0000002           IEA        GO  computed  \n",
       "3732         NaN   PMID:26783301           IMP        GO   curated  \n",
       "3733         NaN   PMID:26783301           IMP        GO   curated  \n",
       "3933         NaN   PMID:21873635           IBA        GO   curated  \n",
       "3935         NaN   PMID:21873635           IBA        GO   curated  \n",
       "6831         NaN   PMID:12828935           IDA        GO   curated  \n",
       "6856         NaN   PMID:12828935           IDA        GO   curated  \n",
       "41065        NaN   PMID:20860624           IDA        GO   curated  \n",
       "41066        NaN   PMID:20860624           IDA        GO   curated  \n",
       "42701        NaN             NaN           NaN  InterPro   curated  \n",
       "42703        NaN             NaN           NaN  InterPro   curated  \n",
       "43236        NaN             NaN           NaN  InterPro   curated  \n",
       "43237        NaN             NaN           NaN  InterPro   curated  \n",
       "43254        NaN             NaN           NaN  InterPro   curated  \n",
       "43255        NaN             NaN           NaN  InterPro   curated  \n",
       "42782        NaN             NaN           NaN  InterPro   curated  \n",
       "42783        NaN             NaN           NaN  InterPro   curated  \n",
       "42806        NaN             NaN           NaN  InterPro   curated  \n",
       "42807        NaN             NaN           NaN  InterPro   curated  \n",
       "42885        NaN             NaN           NaN  InterPro   curated  \n",
       "42886        NaN             NaN           NaN  InterPro   curated  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dup_types = reg_edges[ix]['type'].unique()\n",
    "\n",
    "dup_dfs = []\n",
    "\n",
    "for dt in dup_types:\n",
    "    dup_dfs.append(reg_edges[ix].query('type == @dt').sort_values(['start_id', 'end_id', 'type']).head(6))\n",
    "    \n",
    "pd.concat(dup_dfs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I see no reason for any of these duplications.... We'll just do a simple 'drop_duplicates' rather than merging any columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "reg_edges = reg_edges.drop_duplicates(subset=['start_id', 'end_id', 'type']).copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "43704"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(reg_edges)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43704\n",
      "43704\n"
     ]
    }
   ],
   "source": [
    "all_node_ids = set(all_nodes['id']) \n",
    "\n",
    "print(len(reg_edges))\n",
    "reg_edges_filt = reg_edges.query('start_id in @all_node_ids and end_id in @all_node_ids')\n",
    "print(len(reg_edges_filt))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_id</th>\n",
       "      <th>end_id</th>\n",
       "      <th>type</th>\n",
       "      <th>parent_ixn</th>\n",
       "      <th>pub_med_ids</th>\n",
       "      <th>organism_id</th>\n",
       "      <th>abbv</th>\n",
       "      <th>source</th>\n",
       "      <th>evidence</th>\n",
       "      <th>direct_evidence</th>\n",
       "      <th>corrected_pvalue</th>\n",
       "      <th>inference_gene_symbol</th>\n",
       "      <th>qualifier</th>\n",
       "      <th>db_reference</th>\n",
       "      <th>evidence_code</th>\n",
       "      <th>with_or_from</th>\n",
       "      <th>date</th>\n",
       "      <th>assigned_by</th>\n",
       "      <th>experiments</th>\n",
       "      <th>support_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MESH:C000121</td>\n",
       "      <td>4313</td>\n",
       "      <td>decreases_activity_CdaG</td>\n",
       "      <td>decreases^activity</td>\n",
       "      <td>25899827</td>\n",
       "      <td>9606</td>\n",
       "      <td>CdaG</td>\n",
       "      <td>CTD</td>\n",
       "      <td>curated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MESH:C000121</td>\n",
       "      <td>4313</td>\n",
       "      <td>decreases_expression_CdeG</td>\n",
       "      <td>decreases^expression</td>\n",
       "      <td>25899827</td>\n",
       "      <td>9606</td>\n",
       "      <td>CdeG</td>\n",
       "      <td>CTD</td>\n",
       "      <td>curated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       start_id end_id                       type            parent_ixn  \\\n",
       "0  MESH:C000121   4313    decreases_activity_CdaG    decreases^activity   \n",
       "1  MESH:C000121   4313  decreases_expression_CdeG  decreases^expression   \n",
       "\n",
       "  pub_med_ids organism_id  abbv source evidence direct_evidence  \\\n",
       "0    25899827        9606  CdaG    CTD  curated             NaN   \n",
       "1    25899827        9606  CdeG    CTD  curated             NaN   \n",
       "\n",
       "  corrected_pvalue inference_gene_symbol qualifier db_reference evidence_code  \\\n",
       "0              NaN                   NaN       NaN          NaN           NaN   \n",
       "1              NaN                   NaN       NaN          NaN           NaN   \n",
       "\n",
       "  with_or_from date assigned_by experiments support_type  \n",
       "0          NaN  NaN         NaN         NaN          NaN  \n",
       "1          NaN  NaN         NaN         NaN          NaN  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_edges = pd.concat([edges, reg_edges_filt], sort=False)\n",
    "all_edges.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "126691"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_edge_ids = all_edges[['start_id', 'end_id']].stack()\n",
    "filt_nodes = all_nodes.query('id in @all_edge_ids')\n",
    "len(filt_nodes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save to Disk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "gt.add_colons(reg_edges).to_csv(out_dir.joinpath('new_edges.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "gt.add_colons(all_nodes, id_name='identifier').to_csv(out_dir.joinpath('nodes_all.csv'), index=False)\n",
    "gt.add_colons(filt_nodes, id_name='identifier').to_csv(out_dir.joinpath('nodes_filt.csv'), index=False)\n",
    "\n",
    "gt.add_colons(all_edges).to_csv(out_dir.joinpath('edges.csv'), index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
